---
title: "Simmental GPSM Analysis "
author: "Troy Rowan"
date: "2020-09-20"
output: 
  html_document:
    code_folding: show
editor_options:
  chunk_output_type: inline
  markdown: 
    wrap: 72
---

```{r, setup, include = FALSE}
set.seed(325333)
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
knitr::opts_knit$set(root.dir = here::here())
library(workflowr)
library(knitr)
library(maps)
library(ggthemes)
library(purrr)
library(factoextra)
library(corrr)
library(ggcorrplot)
library(factoextra)
library(here)
library(tidyverse)
library(lubridate)
library(pedigree)
library(qvalue)
library(GALLO)
library(cowplot)
library(gprofiler2)
library(DT)
library(UpSetR)
#source("../code/annotation_functions.R")
```

```{r}
source("code/GCTA_functions.R")
source("code/annotation_functions.R")
simmental = read_csv("output/200907_SIM/phenotypes/200907_SIM.info.csv") %>%
  mutate(ASA_nbr = asa_nbr,
         dummyid = international_id) %>% 
  left_join(read_csv("data/200907_SIM/breeds.csv.gz") %>% 
  pivot_wider(names_from = brd_cd, values_from = percent) %>% 
    replace(is.na(.), 0))

colnames(simmental)
```

# PCA
PC1 eval = 
3073.63
2622.85
777.699
650.589
```{r}
read_delim("output/200907_SIM/pca/200907_SIM.850K.eigenvec", delim = " ",
           col_names = c("FID", "international_id", paste0("PC", seq(1,20)))) %>% 
  left_join(select(simmental, international_id, birth_year, SM), by = "international_id") %>% 
  filter(!is.na(SM)) %>% 
  ggplot(aes(x = PC1, y = PC2, color = SM))+
  geom_point()+
  scale_color_viridis_c()+
  theme_cowplot()

read_delim("output/200907_SIM/pca/200907_SIM.850K.eigenvec", delim = " ",
           col_names = c("FID", "international_id", paste0("PC", seq(1,20)))) %>% 
  left_join(select(simmental, international_id, birth_year, SM, AN), by = "international_id") %>% View()
  filter(!is.na(SM)) %>% 
  ggplot(aes(x = PC1, y = PC2, color = AN))+
  geom_point()+
  scale_color_viridis_c()+
  theme_cowplot()
```


# Simmental GPSM Analysis

## REML variance component estimates {.tabset}

# Purebred Simmental GPSM Analysis

## REML variance component estimates {.tabset}

| Phenotype            	| n     	| h^2   	| SE    	|
|----------------------	|-------	|-------	|-------	|
| Full Age             	| 78787 	| 0.619 	| 0.005   |
| Full Log Age         	| 78787 	| 0.600 	| 0.005   |
| Young Age            	| 73811 	| 0.540   | 0.005   |
| Old Age              	| 4976  	| 0.436 	| 0.021 	|
| SimAngus (AN) Age    	| 11429 	| 0.665 	| 0.011 	|
| SimAngus (SIM) Age   	| 46136 	| 0.642 	| 0.006 	|
| Majority SIM Age     	| 31225 	| 0.558 	| 0.008 	|
| Majority SIM Log Age 	| 31225 	| 0.561 	| 0.008 	|
| Purebred Age         	| 13379 	| 0.555 	| 0.011 	|
| Purebred Log Age     	| 13379 	| 0.560 	| 0.011 	|
| Purebred Young Age   	| 11148 	| 0.497 	| 0.013 	|
| Purebred Old Age     	| 2231  	| 0.462 	| 0.030 	|

## Individual Residuals and Breeding Values {.tabset}

These are REML estimates of individual's breeding values and residuals
from GCTA GREML analysis

### Full Age

All animals with SIM > 0.05

n = 78,787

```{r}
plot_grid(
  read_blp("output/200907_SIM/greml/200907_SIM.full_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Raw Age GPSM\nResiduals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.full_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nBreeding Values")+
    theme_cowplot())

```

### Full Log Age

All animals with SIM > 0.05 

Log-transformed age as dependent variable 

n = 78,787

```{r}
plot_grid(
  read_blp("output/200907_SIM/greml/200907_SIM.full_log_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Log Transformed Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.full_log_age.850K.indi.blp") %>% 
  left_join(simmental %>% 
              select(international_id, age)) %>% 
  ggplot(aes(sample = BV))+
  stat_qq()+
  stat_qq_line(color = "red")+
  ggtitle("\nBreeding Values")+
  theme_cowplot())
```

### Young Age

Animals Born since 2008 with SIM > 0.05

n = 73,811

```{r}
plot_grid(
  read_blp("output/200907_SIM/greml/200907_SIM.full_young_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Young Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.full_young_age.850K.indi.blp") %>% 
  left_join(simmental %>% 
              select(international_id, age)) %>% 
  ggplot(aes(sample = BV))+
  stat_qq()+
  stat_qq_line(color = "red")+
  ggtitle("\nGPSM Breeding Values")+
  theme_cowplot())
```

### Old Age

Animals Born prior to 2008 with SIM > 0.05

n = 4,976

```{r}

plot_grid(
  read_blp("output/200907_SIM/greml/200907_SIM.pb_old_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Old Animals\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.pb_old_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Residuals")+
    theme_cowplot())
```

### SimAngus (Angus) Age

Animals with SIM < 0.30 and ANG > 0.50

```{r}

plot_grid(read_blp("output/200907_SIM/greml/200907_SIM.simangus3050.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("SimAngus (>50% AN)\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.simangus3050.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```

### SimAngus Age

Animals with SIM > 0.20 and SIM < 0.70

n = 11,429

```{r}

plot_grid(read_blp("output/200907_SIM/greml/200907_SIM.simangus2070.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("SimAngus\nRaw Age\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.simangus2070.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```


### Majority Simmental Age

Animals with SIM > 0.70

n = 31,225

```{r}

plot_grid(read_blp("output/200907_SIM/greml/200907_SIM.sim70_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Majority Simmental Animals\nRaw Age\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.sim70_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```


### Majority Simmental Log Age

Animals with SIM > 0.70

n = 31,225

```{r}

plot_grid(read_blp("output/200907_SIM/greml/200907_SIM.sim70_log_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Majority Simmental Animals\nLog Transformed Age\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.sim70_log_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```

### Purebred Age

Animals with SIM = 1.0

n = 13,379

```{r}

plot_grid(read_blp("output/200907_SIM/greml/200907_SIM.pb_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Purebred Simmental\nRaw Age\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.pb_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```

### Purebred Log Age

Animals with SIM = 1.0

n = 13,379

```{r}

plot_grid(read_blp("output/200907_SIM/greml/200907_SIM.pb_log_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Purebred Simmental\nLog Transformed Age\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.pb_log_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```

### Purebred Young Age

Animals with SIM = 1.0 born since 2008

n = 11,148

```{r}

plot_grid(read_blp("output/200907_SIM/greml/200907_SIM.pb_young_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Purebred Simmental (Post-2007)\nRaw Age\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.pb_young_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```

### Purebred Old Age

Animals with SIM = 1, born before 2008

n = 2,231

```{r}

plot_grid(read_blp("output/200907_SIM/greml/200907_SIM.pb_old_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Purebred Simmental (Pre 2008)\nRaw Age\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200907_SIM/greml/200907_SIM.pb_old_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())

read_blp("output/200907_SIM/greml/200907_SIM.pb_old_age.850K.indi.blp") %>% 
    left_join(simmental %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    geom_qq(col = age)+
    stat_qq_line(color = "red")+
    ggtitle("Purebred Simmental (Pre 2008)\nRaw Age\nGPSM Residuals")+
    theme_cowplot()
```

### Transform GPSM Results (Chromosome 1)

```{r}
sim_age =
  read_gwas("output/200907_SIM/gwas/single_chrom/200907_SIM.age.chr1.850K.mlma")

sim_bc = 
  read_gwas("output/200907_SIM/gwas/single_chrom/200907_SIM.bc_age.chr1.850K.mlma")

sim_log = 
  read_gwas("output/200907_SIM/gwas/single_chrom/200907_SIM.log_age.chr1.850K.mlma")

plot_grid(ggmanhattan2(sim_age),
ggmanhattan2(sim_log),
ggmanhattan2(sim_bc),
nrow = 3)

```



## n(SigSNPs)

GWAS results for the 685,120 SNPs with MAF > 0.01 in our imputed dataset. 

```{r eval=TRUE, echo=FALSE}
full_age =
  read_gwas("output/200907_SIM/gwas/200907_SIM.full_age.850K.mlma.gz")

full_log_age =
  read_gwas("output/200907_SIM/gwas/200907_SIM.full_log_age.850K.mlma.gz")

young_age =
  read_gwas("output/200907_SIM/gwas/200907_SIM.full_young_age.850K.mlma.gz")

old_age = 
  read_gwas("output/200907_SIM/gwas/200907_SIM.full_old_age.850K.mlma.gz")

simangus_an_age = 
  read_gwas("output/200907_SIM/gwas/200907_SIM.simangus3050.850K.mlma.gz")

simangus_age = 
  read_gwas("output/200907_SIM/gwas/200907_SIM.simangus2070.850K.mlma.gz")

maj_sim_age = 
  read_gwas("output/200907_SIM/gwas/200907_SIM.sim70_age.850K.mlma.gz")

maj_sim_log_age = 
  read_gwas("output/200907_SIM/gwas/200907_SIM.sim70_log_age.850K.mlma.gz")

pb_age = 
  read_gwas("output/200907_SIM/gwas/200907_SIM.pb_age.850K.mlma.gz")

pb_log_age = 
  read_gwas("output/200907_SIM/gwas/200907_SIM.pb_log_age.850K.mlma.gz")

pb_young_age = 
  read_gwas("output/200907_SIM/gwas/200907_SIM.pb_young_age.850K.mlma.gz")

pb_old_age = 
  read_gwas("output/200907_SIM/gwas/200907_SIM.pb_old_age.850K.mlma.gz")

sigcheck = function(df){

  gwasdf = 
    df %>% 
    mutate(Nominal = case_when(p <= 1e-5 ~ TRUE,
                             TRUE ~ FALSE),
           Bonferroni = case_when(p <= 7.55e-07 ~ TRUE,
                             TRUE ~ FALSE),
           q0.1 = case_when(q <= 0.10 ~ TRUE,
                             TRUE ~ FALSE),
           q0.05 = case_when(q <= 0.05 ~ TRUE,
                             TRUE ~ FALSE))%>% 
  select(Nominal, Bonferroni, q0.1, q0.05) %>% 
  colSums()
  return(gwasdf)
}

rbind(sigcheck(full_age),
      sigcheck(full_log_age),
      sigcheck(young_age),
      sigcheck(old_age),
      sigcheck(simangus_an_age),
      sigcheck(simangus_age),
      sigcheck(maj_sim_age),
      sigcheck(maj_sim_log_age),
      sigcheck(pb_age),
      sigcheck(pb_young_age),
      sigcheck(pb_old_age)) %>% as.data.frame() %>% 
mutate(analysis = c("full_age", "full_log_age", "young_age","old_age", "simangus_an_age", "simangus_age", "maj_sim_age", "maj_sim_log_age", "pb_age", "pb_young_age", "pb_old_age")) %>% 
select(analysis, everything()) %>% 
datatable(caption = "Number of significant SNPs at various significance levels with different dependent variables")
  
  
```

## GPSM GWAS Manhattan Plots for Transformed Age {.tabset}

### Full Age

All animals with SIM > 0.05

n = 78,787

```{r}
ggmanhattan2(full_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(full_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
```

### Full Log Age

All animals with SIM > 0.05 

Log-transformed age as dependent variable 

n = 78,787

```{r}
ggmanhattan2(full_log_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(full_log_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
```

### Young Age

Animals Born since 2008 with SIM > 0.05

n = 73,811

```{r}
ggmanhattan2(young_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(young_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
```

### Old Age

Animals Born prior to 2008 with SIM > 0.05

n = 4,976

```{r}
ggmanhattan2(old_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(old_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
```

### SimAngus (Angus) Age

Animals with SIM < 0.30 and ANG > 0.50

```{r}
ggmanhattan2(simangus_an_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(simangus_an_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
```

### SimAngus Age

Animals with SIM > 0.20 and SIM < 0.70

n = 11,429

```{r}
ggmanhattan2(simangus_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(simangus_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
```


### Majority Simmental Age

Animals with SIM > 0.70

n = 31,225

```{r}
ggmanhattan2(maj_sim_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(maj_sim_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
```


### Majority Simmental Log Age

Animals with SIM > 0.70

n = 31,225

```{r}
ggmanhattan2(maj_sim_log_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(maj_sim_log_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
```

### Purebred Age

Animals with SIM = 1.0

n = 13,379

```{r}
ggmanhattan2(pb_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(pb_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
```

### Purebred Log Age

Animals with SIM = 1.0

n = 13,379

```{r}
ggmanhattan2(pb_log_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(pb_log_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
```

### Purebred Young Age

Animals with SIM = 1.0 born since 2008

n = 11,148

```{r}

ggmanhattan2(pb_young_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(pb_young_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
```

### Purebred Old Age

Animals with SIM = 1, born before 2008

n = 2,231

```{r}

ggmanhattan2(pb_old_age,
               prune = 0.01, 
               sig_threshold_p = 7.298e-08)

ggmanhattan2(pb_old_age,
             value = q,
             prune = 0.9,
             sig_threshold_q = 0.1)
  
```

## Sequence-imputed GPSM

Used variants with MAF > 0.01 and imputation R^2 values > 0.4

During reading in I filtered p < 0.01 to make sure that R is able to handle these large files without crashing.

our Bonferroni multiple-testing cutoff is 0.05/11657021 4.289e-09

In all cases henceforth, I'm working with p-values. Calculating q-values needs the genome-wide p-values, and I'm filtering here to prevent computer from crashing. Easy enough to do this for q-values downstream though i suppose

```{r}
sim_seq_gpsm_age = 
  seq(1,29) %>% 
    purrr::map(~read_gwas(paste0("output/200907_SIM/seq_gwas/200907_SIM.full_age.chr",.x,".mlma.gz")) %>% 
                filter(p < 0.01)) %>% 
    purrr::reduce(bind_rows)

sim_seq_gpsm_purebred_age = 
  seq(1,29) %>% 
    purrr::map(~read_gwas(paste0("output/200907_SIM/seq_gwas/200907_SIM.pb_age.chr",.x,".mlma.gz")) %>% 
                 filter(p < 0.01)) %>% 
    purrr::reduce(bind_rows)

sim_seq_gpsm_purebred_sigsnps =
  sim_seq_gpsm_purebred_age %>% 
  filter(p < 4.289e-9) %>% 
  .$SNP

sim_seq_gpsm_sigsnps =
  sim_seq_gpsm_age %>% 
  filter(p < 4.289e-9) %>% 
  .$SNP

sim_seq_gpsm_purebred_age %>% 
  top_n(1000, wt = -p) %>% 
  View()
```


### Manhattan Plots {.tabset}

#### Purebred Simmental Manhattan Plot

```{r}
sim_seq_gpsm_purebred_age %>% 
  ggmanhattan2(sig_threshold_p = 4.289e-9,
               sigsnps = sim_pb_seq_cojo$SNP)+
  ggtitle("Purebred Simmental Age GPSM")+
  geom_hline(yintercept = 7.3, color = "blue")
```

#### Full  Simmental Manhattan Plot
```{r}
sim_seq_gpsm_age %>% 
  ggmanhattan2(sig_threshold_p = 4.289e-9,
               sigsnps = sim_seq_cojo$SNP)+
  ggtitle("Full Simmental GPSM")+
  geom_hline(yintercept = 7.3, color = "blue")
```

#### Single Chromosome (Still RAN, copied code)

```{r}
single_chrom = read_gwas("output/200910_RAN/seq_gwas/200910_RAN.age.chr6.mlma.gz") %>% 
  mutate(COJO = case_when(SNP %in% ran_age_seq_cojo$SNP ~ TRUE,
                          TRUE ~ FALSE),
         BP = BP/1000000)


plot_grid(
    single_chrom %>% 
      # filter(BP > 8e7,
      #        BP <10e7) %>% 
    ggmanhattan2(prune = 0.1,
                 sigsnps = single_chrom %>% 
                   filter(COJO == TRUE) %>% 
                   .$SNP,
                 sig_threshold_p = 5e-8)+
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10))+
  labs(x = "Megabases"),
  ran_nsl %>% 
    filter(CHR == 6) %>% 
    selscan_manhattans(col = abs(norm_nsl))+
    geom_vline(xintercept = single_chrom %>% 
                   filter(COJO == TRUE) %>% 
                   .$BP, 
               color = "springgreen3",
               alpha = 0.5)+
    labs(y = "Normalized nSL", x = "Megabases")+
    geom_hline(yintercept = 3, color = "red")+
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)),
  ran_raisd %>% 
    mutate(BP = BP/1000000) %>% 
    filter(CHR == 6,
           Mu > 50) %>% 
    selscan_manhattans(col = Mu)+
    geom_vline(xintercept = single_chrom %>% 
                   filter(COJO == TRUE) %>% 
                   .$BP, 
               color = "springgreen3",
               alpha = 0.5)+
    labs(y = "RAiSD Mu", x = "Megabases")+
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)),
  nrow = 3)

ran_nsl =
  read_tsv("output/200910_RAN/selscan/nsl/200910_RAN.all.nsl.out.100bins.norm.gz",
             col_names = c("SNP", "BP", "freq", "sl1", "sl0", "nsl", "norm_nsl", "sig")) %>% 
    mutate(CHR = as.numeric(str_split_fixed(SNP, ":", n = 4)[,1]),
           SNP = paste(CHR, BP, sep = ":"),
         sig = case_when(abs(norm_nsl) > 3 ~ TRUE,
                         TRUE ~ FALSE),
         BP = BP/1000000)


single_chrom_manhattan(single_chrom, prune = 1)
  
single_chrom_manhattan(prune = 1)


ran_nsl_window = 
  read_tsv("output/200910_RAN/selscan/nsl/200910_RAN.all.nsl.windows.txt") %>% 
  mutate(nsl_win = abs(Wstat),
         SNP = paste(CHR, BP, sep = ":")) 
  
```


## Sequence-imputed COJO


[COJO documentation](https://cnsgenomics.com/software/gcta/#COJO)

[COJO citation](https://www.nature.com/articles/ng.2213)

Used `--cojo-slct` procedure with a significance cutoff of 5e-8

Reading in single-chromosome COJO files.
```{r}
sim_pb_seq_cojo = 
  list.files(path='output/200907_SIM/seq_cojo/',pattern="200907_SIM.pb_age.chr") %>% 
  purrr::map(
  ~read_tsv(paste0("output/200907_SIM/seq_cojo/",.x),
           col_names = c("Chr", "SNP", "bp", "freq", "refA", "b", "se", "p", "n", "freq_geno", "bJ", "bJ_se", "pJ", "LD_r"),
           skip = 1)  %>% 
    mutate(chrbp = paste(Chr, bp, sep = ":")) %>% 
    select(SNP, CHR = Chr, BP = bp, p, pJ, chrbp)) %>% 
    purrr::reduce(bind_rows) 

sim_pb_seq_cojo %>% 
  select(SNP) %>% 
  write_tsv("output/200907_SIM/seq_cojo/200907_SIM.purebred.cojosnps.txt",
            col_names = FALSE)


sim_pb_seq_cojo_genes =
  sim_pb_seq_cojo %>% 
    gene_annotation() %>% 
      left_join(sim_pb_seq_cojo, 
                by = c("CHR", "BP")) %>% 
    select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% 
  left_join(sim_pb_seq_cojo)


sim_seq_cojo = 
  list.files(path='output/200907_SIM/seq_cojo/',pattern="200907_SIM.full_age") %>% 
  purrr::map(
  ~read_tsv(paste0("output/200907_SIM/seq_cojo/",.x),
           col_names = c("Chr", "SNP", "bp", "freq", "refA", "b", "se", "p", "n", "freq_geno", "bJ", "bJ_se", "pJ", "LD_r"),
           skip = 1)  %>% 
    mutate(chrbp = paste(Chr, bp, sep = ":")) %>% 
    select(SNP, CHR = Chr, BP = bp, p, pJ, chrbp)) %>% 
    purrr::reduce(bind_rows) 

sim_seq_cojo %>% 
  select(SNP) %>% 
  write_tsv("output/200907_SIM/seq_cojo/200907_SIM.full.cojosnps.txt",
            col_names = FALSE)


sim_seq_cojo_genes =
  sim_seq_cojo %>% 
    gene_annotation() %>% 
      left_join(sim_seq_cojo, 
                by = c("CHR", "BP")) %>% 
    select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream)
filter(sim_seq_cojo,
       SNP %in% sim_pb_seq_cojo$SNP)


sim_pb_seq_cojo %>% 
  gene_annotation() %>% 
    right_join(sim_pb_seq_cojo, 
              by = c("CHR", "BP")) %>% 
  select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% write_tsv("output/200907_SIM/genes/200907_SIM.pb_age.gpsm.cojogenes.50kb.txt") 
```

### Gene overlap with other sweep regions

So that's 2 of 34 genes in PB Simmental GPSM are also identified by Sweep Regions. 
```{r}


map2(.x = sim_seq_cojo$CHR, .y = sim_seq_cojo$BP,
  ~filter(
    sim_raisd,
    Mu > 193.5,
    CHR == .x,
    stat <= .y,
    finish >= .y) %>% 
    mutate(COJO_SNP = paste(.x, .y, sep = ":"))) %>% 
  reduce(bind_rows) 


map2(.x = sim_pb_seq_cojo$CHR, .y = sim_pb_seq_cojo$BP,
  ~filter(
    sim_raisd,
    Mu > 193.5,
    CHR == .x,
    stat <= .y,
    finish >= .y)) %>% 
  reduce(bind_rows) 

map2(.x = sim_seq_cojo$CHR, .y = sim_seq_cojo$BP,
  ~filter(
    sim_nsl,
    MeanY > 2.381382,
    CHR == .x,
    WindowStart <= .y,
    WindowStop >= .y)) %>% 
  reduce(bind_rows) 

map2(.x = sim_pb_seq_cojo$CHR, .y = sim_pb_seq_cojo$BP,
  ~filter(
    sim_nsl,
    MeanY > 2.381382,
    CHR == .x,
    WindowStart <= .y,
    WindowStop >= .y)) %>% 
  reduce(bind_rows) 

```



### Gene Annotations for COJO hits:

Annotated genes within 50 kb of COJO SNP

#### Purebred
```{r}
sim_pb_seq_cojo_genes %>%
  datatable()

sim_pb_seq_cojo_genes %>% 
  mutate(ingene = case_when(distance == 0 ~ "within",
                            is.na(distance) ~ ">50kb",
                            TRUE ~ "proximal")) %>% 
  group_by(SNP) %>% 
  top_n(1, wt = -distance) %>% 
  ungroup() %>% 
  count(ingene)
```

#### Full Dataset
```{r}
sim_seq_cojo_genes %>% write_csv("../analysis/genes.txt")

sim_seq_cojo_genes %>% 
  mutate(ingene = case_when(distance == 0 ~ "within",
                            is.na(distance) ~ ">50kb",
                            TRUE ~ "proximal")) %>% 
  group_by(SNP) %>% 
  top_n(1, wt = -distance) %>% 
  ungroup() %>% 
  count(ingene)
```

### Gene Ontology Analysis

Gene ontology enrichment analysis using `gprofiler2::gost()` of gene names from GALLO annotation of log-transformed GPSM COJO hits

Significance threshold set at FDR < 0.10

####Purebred
```{r, eval }
sim_pb_seq_cojo_GO =
  sim_pb_seq_cojo_genes %>% 
  .$gene_id %>% unique() %>% 
  gost(query = ., 
      organism = "btaurus",
      user_threshold = 0.1,
      correction_method = "fdr")
sim_pb_seq_cojo_GO$result


gostplot(sim_pb_seq_cojo_GO, capped = TRUE, interactive = TRUE)
```

#### Full Dataset
```{r, eval }
sim_seq_cojo_GO =
  sim_seq_cojo_genes %>% 
  .$gene_id %>% unique() %>% 
  gost(query = ., 
      organism = "btaurus",
      user_threshold = 0.1,
      correction_method = "fdr")


gostplot(sim_seq_cojo_GO, capped = TRUE, interactive = TRUE)
```

### QTL Enrichments {.tabset}


Removed milk-associated QTL, perform enrichment on whole genome QTL and by chromosome

#### Purebred (Genome)
```{r}
sim_pb_seq_cojo %>%
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```

#### Full Dataset (Genome)
```{r}
sim_seq_cojo %>%
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```

#### purebred (per-Chromosome)
```{r}
sim_pb_seq_cojo %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```

#### Young (per-chromosme)
```{r}
ran_young_seq_cojo %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```

### FAETH Scores of Significant loci
13 non-NA FAETH SNPs in Purebred (6 in top 1/3)
177 non-NA FAETH SNPs in Full (69)


```{r}
# sim_pb_FAETH =
#   left_join(sim_pb_seq_cojo, FAETH_Scores, by = c("chrbp" = "SNP"))
# 
# sim_full_FAETH =
#   left_join(sim_seq_cojo, FAETH_Scores, by = c("chrbp" = "SNP")) %>% 
#   filter(!is.na(FAETH))

sim_full_FAETH %>% 
  filter(!is.na(FAETH),
         FAETH > 1.6e-8)
```


### Annnotating with FAANG data

```{r}
sim_seq_cojo_functional = 
  map2(.x = sim_seq_cojo$CHR, .y = sim_seq_cojo$BP,
    ~filter(
      FAANG,
      CHR == .x,
      START <= .y,
      STOP >= .y)%>% 
      mutate(COJO_SNP = paste(.x, .y, sep = ":"))) %>% 
    reduce(bind_rows) %>% 
  left_join(sim_seq_cojo,
            by = c("COJO_SNP" = "chrbp")) 


pb_sim_seq_cojo_functional = 
  map2(.x = sim_pb_seq_cojo$CHR, .y = sim_pb_seq_cojo$BP,
    ~filter(
      FAANG,
      CHR == .x,
      START <= .y,
      STOP >= .y)%>% 
      mutate(COJO_SNP = paste(.x, .y, sep = ":"))) %>% 
    reduce(bind_rows) %>% 
  left_join(sim_pb_seq_cojo, 
            by = c("COJO_SNP" = "chrbp"))



pb_sim_seq_cojo_functional %>% 
  count(COJO_SNP)


sim_seq_cojo_functional %>% 
  select(SNP,TISSUE) %>% 
  distinct() %>% 
  mutate(case = as.integer(1)) %>% 
  pivot_wider(names_from = TISSUE, values_from = case, values_fill = as.integer(0)) %>% 
  as.data.frame() %>% 
  upset(nsets = 8, order.by = "freq")

intersect(sim_seq_cojo$SNP, ran_young_seq_cojo$SNP)

upset(movies, nsets = 6, number.angles = 30, point.size = 3.5, line.size = 2, 
    mainbar.y.label = "Genre Intersections", sets.x.label = "Movies Per Genre", 
    text.scale = c(1.3, 1.3, 1, 1, 2, 0.75))
```

### Gene overlap with nSL regions


```{r}

map2(.x = sim_pb_seq_cojo$CHR, .y = sim_pb_seq_cojo$BP,
  ~filter(
    pb_sim_nsl_sig_windows,
    CHR == .x,
    WindowStart-100000 <= .y,
    WindowStop+100000 >= .y)) %>% 
  reduce(bind_rows) 


```
