---
title: "GPSM"
author: "Troy Rowan"
date: "2020-09-20"
output: 
  html_document:
    code_folding: show
editor_options:
  chunk_output_type: inline
  markdown: 
    wrap: 72
---

```{r, setup, include = FALSE}
set.seed(325333)
library(workflowr)
library(knitr)
library(maps)
library(ggthemes)
library(purrr)
library(factoextra)
library(corrr)
library(ggcorrplot)
library(factoextra)
library(here)
library(tidyverse)
library(lubridate)
library(pedigree)
library(qvalue)
library(GALLO)
library(cowplot)
library(gprofiler2)
library(DT)
library(ggforce)
library(vroom)
library(qvalue)
library(moments)
#source("../code/annotation_functions.R")
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
source("code/GCTA_functions.R")
source("code/annotation_functions.R")
simmental = read_csv("output/200907_SIM/phenotypes/200907_SIM.info.csv")
redangus = read_csv("output/200910_RAN/phenotypes/200910_RAN.info.csv")
```



# Red Angus GPSM Analysis

## REML variance component estimates {.tabset}

### Raw Age

Raw Age (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 5.244    | 0.128 |
| V(e)    | 4.789    | 0.040 |
| V(p)    | 10.03    | 0.118 |
| V(G)/Vp | 0.523    | 0.007 |

### Square Root

Square Root Transformed Age (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.242    | 0.005 |
| V(e)    | 0.151    | 0.001 |
| V(p)    | 0.393    | 0.005 |
| V(G)/Vp | 0.616    | 0.006 |

### Cube Root

Cube Root Transformed Age (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.065    | 0.001 |
| V(e)    | 0.037    | 0.000 |
| V(p)    | 0.101    | 0.001 |
| V(G)/Vp | 0.637    | 0.006 |

### Box-Cox

Box-Cox Transformed Age (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.115    | 0.002 |
| V(e)    | 0.060    | 0.001 |
| V(p)    | 0.175    | 0.002 |
| V(G)/Vp | 0.657    | 0.006 |

### Log

Log Transformed Age (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.222    | 0.005 |
| V(e)    | 0.115    | 0.001 |
| V(p)    | 0.337    | 0.005 |
| V(G)/Vp | 0.657    | 0.006 |

### Young Raw Age 

Raw Age (Born 2012-Present) (n= 44,470):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 3.354    | 0.061 |
| V(e)    | 0.983    | 0.001 |
| V(p)    | 4.337    | 0.057 |
| V(G)/Vp | 0.773    | 0.004 |

### Young Log Age

Log Transformed Age (Born 2012-Present) (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.200    | 0.004 |
| V(e)    | 0.079    | 0.001 |
| V(p)    | 0.279    | 0.004 |
| V(G)/Vp | 0.718    | 0.005 |

### Old Raw Age 

Raw Age (Born 2012-Present) (n= 1,984):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 16.70    | 1.277 |
| V(e)    | 13.30    | 0.854 |
| V(p)    | 30.00    | 1.041 |
| V(G)/Vp | 0.557    | 0.031 |

### Old Log Age

Log Transformed Age (Born 2012-Present) (n= 1,984):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.056    | 0.004 |
| V(e)    | 0.045    | 0.003 |
| V(p)    | 0.100    | 0.003 |
| V(G)/Vp | 0.555    | 0.031 |

## Individual Residuals and Breeding Values {.tabset}

These are REML estimates of individual's breeding values and residuals
from GCTA GREML analysis

### Age

```{r}
plot_grid(
  read_blp("output/200910_RAN/greml/200910_RAN.age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Raw Age GPSM\nResiduals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nBreeding Values")+
    theme_cowplot())
```

### Square Root

```{r}
plot_grid(
  read_blp("output/200910_RAN/greml/200910_RAN.sqrt_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Square Root Transformed Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.sqrt_age.850K.indi.blp") %>% 
  left_join(redangus %>% 
              select(international_id, age)) %>% 
  ggplot(aes(sample = BV))+
  stat_qq()+
  stat_qq_line(color = "red")+
  ggtitle("\nBreeding Values")+
  theme_cowplot())
```

### Cube Root

```{r}
plot_grid(
  read_blp("output/200910_RAN/greml/200910_RAN.cbrt_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Cube Root Transformed Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.cbrt_age.850K.indi.blp") %>% 
  left_join(redangus %>% 
              select(international_id, age)) %>% 
  ggplot(aes(sample = BV))+
  stat_qq()+
  stat_qq_line(color = "red")+
  ggtitle("\nGPSM Breeding Values")+
  theme_cowplot())
```

### Box-Cox

```{r}

plot_grid(
  read_blp("output/200910_RAN/greml/200910_RAN.bc_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Box-Cox Transformed Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.bc_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Residuals")+
    theme_cowplot())
```

### Log

```{r}

plot_grid(read_blp("output/200910_RAN/greml/200910_RAN.log_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Log Transformed Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.log_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```

### Youg Raw Age

```{r}

plot_grid(read_blp("output/200910_RAN/greml/200910_RAN.young_age.850K.indi.blp.gz") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("(2012-Present) Animals Only\n Raw Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.young_age.850K.indi.blp.gz") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```


### Log

```{r}

plot_grid(read_blp("output/200910_RAN/greml/200910_RAN.young_log_age.850K.indi.blp.gz") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Log Transformed Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.young_log_age.850K.indi.blp.gz") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```


### Old Raw Age

```{r}

plot_grid(read_blp("output/200910_RAN/greml/200910_RAN.old_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("(2011 and Earlier)\n Raw Age\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.old_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())
```

### Old log Age

```{r}

plot_grid(read_blp("output/200910_RAN/greml/200910_RAN.old_log_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("(2011 and Earlier)\n Raw Age\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.old_log_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```


## Breeding Value Correlations Across Transforms/SUbsets  

```{r}
blup_compare =
  read_blp("output/200910_RAN/greml/200910_RAN.young_age.850K.indi.blp.gz") %>% 
    left_join(read_blp("output/200910_RAN/greml/200910_RAN.log_age.850K.indi.blp") %>% 
                select(international_id, logageBV = BV, logageResidual = Residual))

cor(blup_compare$BV, blup_compare$logageBV)

old_blups =
  read_blp("output/200910_RAN/greml/200910_RAN.age.850K.indi.blp") %>%
  filter(international_id %in% read_blp("output/200910_RAN/greml/200910_RAN.old_age.850K.indi.blp")$international_id) %>% 
    left_join(read_blp("output/200910_RAN/greml/200910_RAN.log_age.850K.indi.blp") %>% 
                select(international_id, logageBV = BV, logageResidual = Residual))

cor(old_blups$BV, old_blups$logageBV)

young_blups =
  read_blp("output/200910_RAN/greml/200910_RAN.age.850K.indi.blp") %>%
  filter(!international_id %in% read_blp("output/200910_RAN/greml/200910_RAN.old_age.850K.indi.blp")$international_id) %>% 
    left_join(read_blp("output/200910_RAN/greml/200910_RAN.log_age.850K.indi.blp") %>% 
                select(international_id, logageBV = BV, logageResidual = Residual))

cor(young_blups$BV, young_blups$logageBV)

```


## n(SigSNPs)

The number of significant SNPs in each analysis at various significance
level cutoffs for both p and q values

|        | p\<1e-5 | p\<1e7.55e-7 | q\<0.1 | q\<0.05 |
|--------|---------|--------------|--------|---------|
| Raw    | 315     | 214          | 509    | 398     |
| Sqrt   | 453     | 333          | 729    | 559     |
| Cbrt   | 471     | 357          | 817    | 596     |
| BoxCox | 540     | 390          | 907    | 754     |
| Log    | 513     | 377          | 822    | 715     |
| YAge   | 762     | 555          | 1210   | 1042    |
| YLog   | 743     | 533          | 1123   | 989     |
| OAge   | 762     | 555          | 1210   | 1042    |
| OLog   | 743     | 533          | 1123   | 989     |

```{r eval=TRUE, echo=FALSE}
sigcheck = function(df){
  gwasdf = 
    df %>% 
    mutate(p_1e5 = case_when(p <= 1e-5 ~ TRUE,
                             TRUE ~ FALSE),
           bonferr = case_when(p <= 7.55e-7 ~ TRUE,
                             TRUE ~ FALSE),
           q01 = case_when(q <= 0.10 ~ TRUE,
                             TRUE ~ FALSE),
           q005 = case_when(q <= 0.05 ~ TRUE,
                             TRUE ~ FALSE))%>% 
  select(p_1e5, bonferr, q01, q005) %>% 
  colSums() 
  return(gwasdf)
}

ran_gpsm_age = 
  read_gwas2("output/200910_RAN/gwas/200910_RAN.age.850K.mlma.gz")

ran_gpsm_sqrtage = 
  read_gwas2("output/200910_RAN/gwas/200910_RAN.sqrt_age.850K.mlma.gz")

ran_gpsm_cbrtage = 
  read_gwas2("output/200910_RAN/gwas/200910_RAN.cbrt_age.850K.mlma.gz")

ran_gpsm_bcage = 
  read_gwas2("output/200910_RAN/gwas/200910_RAN.bc_age.850K.mlma.gz")

ran_gpsm_logage = 
  read_gwas2("output/200910_RAN/gwas/200910_RAN.log_age.850K.mlma.gz")

ran_gpsm_youngage = 
  read_gwas2("output/200910_RAN/gwas/200910_RAN.young_age.850K.mlma.gz")

ran_gpsm_younglogage = 
  read_gwas2("output/200910_RAN/gwas/200910_RAN.young_log_age.850K.mlma.gz")

ran_gpsm_oldage = 
  read_gwas2("output/200910_RAN/gwas/200910_RAN.old_age.850K.mlma.gz")

ran_gpsm_oldlogage = 
  read_gwas("output/200910_RAN/gwas/200910_RAN.old_log_age.850K.mlma.gz")

rbind(sigcheck(ran_gpsm_age),
      sigcheck(ran_gpsm_sqrtage),
      sigcheck(ran_gpsm_cbrtage),
      sigcheck(ran_gpsm_bcage),
      sigcheck(ran_gpsm_logage),
      sigcheck(ran_gpsm_youngage),
      sigcheck(ran_gpsm_younglogage),
      sigcheck(ran_gpsm_oldage),
      sigcheck(ran_gpsm_oldlogage)) %>% as.data.frame() %>% 
mutate(analysis = c("ran_gpsm_age", "ran_gpsm_sqrtage", "ran_gpsm_cbrtage","ran_gpsm_bcage", "ran_gpsm_logage", "ran_gpsm_youngage", "ran_gpsm_oldage", "ran_gpsm_oldlogage")) %>% 
select(analysis, everything()) %>% 
datatable(caption = "Number of significant SNPs at various significance levels with different dependent variables")
  
```


### Checking normality of GPSM betas
```{r}
skewness(ran_gpsm_age$b)
skewness(ran_gpsm_logage$b)
skewness(ran_gpsm_bcage$b)

skewness(redangus$age)

ran_gpsm_age %>% 
  ggplot(aes(b))+
  geom_histogram()


skewness(rnorm(n = 800000, mean = 0, sd = 0.051))


```


## GPSM GWAS Manhattan Plots for Transformed Age {.tabset}

### Raw Age

(Significance threshold - Bonferroni)

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_age,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07),
  ggmanhattan2(ran_gpsm_age,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07)+
    ylim(c(0,15)),
  nrow = 2)

#Saving significant SNPs for highlighting in other plots:
raw_age_sigsnps = 
  ran_gpsm_age %>% 
    filter(p < 7.546167e-07) %>% .$SNP
```

### Square Root

Square root transformed age as phenotype (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_sqrtage,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_sqrtage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_sqrtage,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_sqrtage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)
```

### Cube Root

Cube Root Transformed Age Manahattan Plots (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_cbrtage,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_cbrtage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_cbrtage,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_cbrtage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)
```

### Box-Cox

Box-Cox Transformed Age Manahattan Plots (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_bcage,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_bcage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_bcage,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_bcage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)
```

### Log

Log Transformed Age Manahattan Plots (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_logage %>% 
                 filter(CHR <= 29),
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_logage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_logage%>% 
                 filter(CHR <= 29),
               prune = 0.1,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_logage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)
ggsave("output/figures/200910_RAN.logage.850K.manhattan.png", width = 12, height = 8)


plot_grid(
  ggmanhattan2(ran_gpsm_logage,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07),
  ggmanhattan2(ran_gpsm_logage,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07)+
    ylim(c(0,15)),
  nrow = 2)

logage_sigsnps = 
  filter(ran_gpsm_logage,
         p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% .$SNP
  
```

### Young Raw Age

Young animal only (Born after Jan 1, 2012) Raw Age Manahattan Plots (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_youngage,
               prune = 0.01, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & SNP %in% c(raw_age_sigsnps, logage_sigsnps)) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_youngage,
               prune = 0.01,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & SNP %in% c(raw_age_sigsnps, logage_sigsnps)) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)

plot_grid(
  ggmanhattan2(ran_gpsm_youngage,
               prune = 0.01, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = logage_sigsnps),
  ggmanhattan2(ran_gpsm_youngage,
               prune = 0.01,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = logage_sigsnps)+
    ylim(c(0,15)),
  nrow = 2)
```

### Old  Age

Old animal only (Born before Jan 1, 2012) Raw Age and log-transformed Manhattan Plots (Significance threshold - 1e-5)

Green points indicate novel SNPs in this suset analysis that weren't identified in the log-transformed GPSM analysis of log-transformed age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_oldage,
             prune = 0.01, 
             sig_threshold_p = 1e-5, 
             sigsnps = logage_sigsnps)+
  labs(title = "Old Animals (2011 and older): Raw Age"),
  ggmanhattan2(ran_gpsm_oldlogage,
               prune = 0.01, 
               sig_threshold_p = 1e-5, 
               sigsnps = logage_sigsnps)+
    labs(title = "Old Animals (2011 and older): Log-transformed Age"),
nrow = 2)
```

### Young Log Age

Young animal only (Born after Jan 1, 2012) Raw Age Manahattan Plots (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_younlogage,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP, 
               sigsnps2 = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & !SNP %in% logage_sigsnps) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_youngage,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP, 
               sigsnps2 = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & !SNP %in% logage_sigsnps) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)
```


## Sequence-imputed GPSM

Used variants with MAF > 0.01 and imputation R^2 values > 0.4

During reading in I filtered p < 0.01 to make sure that R is able to handle these large files without crashing.

our Bonferroni multiple-testing cutoff is 0.05/11657021 4.289e-09

In all cases henceforth, I'm working with p-values. Calculating q-values needs the genome-wide p-values, and I'm filtering here to prevent computer from crashing. Easy enough to do this for q-values downstream though i suppose

```{r}
# ran_seq_gpsm_logage = 
#   seq(1,29) %>% 
#     purrr::map(~read_gwas(paste0("output/200910_RAN/seq_gwas/200910_RAN.log_age.chr",.x,".mlma.gz")) %>% 
#                 filter(p < 0.01)) %>% 
#     purrr::reduce(bind_rows)

ran_seq_gpsm_age = #read_csv("output/200910_RAN/seq_gwas/200910_RAN.age.combined.mlma.gz")
  seq(1,29) %>%
    purrr::map(~read_gwas(paste0("output/200910_RAN/seq_gwas/200910_RAN.age.chr",.x,".mlma.gz")) %>% 
                 filter(p<0.01)) %>%
    purrr::reduce(bind_rows)
  
# write_csv(ran_seq_gpsm_age, "output/200910_RAN/seq_gwas/200910_RAN.age.combined.mlma.gz")

ran_gpsm_sigsnps =
  ran_seq_gpsm_age %>% 
  filter(q < 0.1) %>% 
  .$SNP

ran_gpsm_sigsnps_bonferroni =
  ran_seq_gpsm_age %>% 
  filter(p < 4.289e-9) %>% 
  .$SNP

ran_seq_gpsm_young_age = #read_csv("output/200910_RAN/seq_gwas/200910_RAN.young_age.combined.mlma.gz")
  seq(1,29) %>%
    purrr::map(~read_gwas(paste0("output/200910_RAN/seq_gwas/200910_RAN.young_age.chr",.x,".mlma.gz")) %>% 
                 filter(p<0.01)) %>%
    purrr::reduce(bind_rows) 
 #   filter(!is.na(p), p > 0) %>% 
#   mutate(q = qvalue(p)$qvalues) %>% 
#   filter(p<0.01)
# write_csv(ran_seq_gpsm_young_age, "output/200910_RAN/seq_gwas/200910_RAN.young_age.combined.mlma.gz")
# ran_seq_gpsm_old_age =
#   seq(1,29) %>%
#     purrr::map(~read_gwas(paste0("output/200910_RAN/seq_gwas/200910_RAN.old_age.chr",.x,".mlma.gz")) %>%
#                  filter(p < 0.01)) %>%
#     purrr::reduce(bind_rows)

ran_seq_gpsm_age %>% 
  filter(p < 4.289e-9) %>% 
  count()

ran_seq_gpsm_age %>% 
  filter(p < 5e-8) %>% 
  count()

ran_seq_gpsm_age %>% 
  filter(q < 0.1) %>% 
  count()

ran_seq_gpsm_young_age %>% 
  filter(p < 5e-8) %>% 
  count()

ran_seq_gpsm_old_age %>% 
  filter(p < 5e-8) %>% 
  count()

```
## Sequence-imputed COJO


[COJO documentation](https://cnsgenomics.com/software/gcta/#COJO)

[COJO citation](https://www.nature.com/articles/ng.2213)

Used `--cojo-slct` procedure with a significance cutoff of 5e-8
For full dataset, raw age:
248 are significant without genome-wide cutoff
103 when only looking at q < 0.1
97 when cutoff is q < 0.05
89 when cutoff is q < 0.01
72 when cutoff is p < 5e-8

Reading in single-chromosome COJO files.
```{r}
ran_age_seq_cojo = 
  list.files(path='output/200910_RAN/seq_cojo/',pattern="200910_RAN.age.chr") %>% 
  purrr::map(
  ~read_tsv(paste0("output/200910_RAN/seq_cojo/",.x),
           col_names = c("Chr", "SNP", "bp", "freq", "refA", "b", "se", "p", "n", "freq_geno", "bJ", "bJ_se", "pJ", "LD_r"),
           skip = 1) %>% 
    mutate(chrbp = paste(Chr, bp, sep = ":")) %>% 
    select(SNP, CHR = Chr, BP = bp, p, pJ, chrbp)) %>% 
    purrr::reduce(bind_rows) %>% 
  filter(p < 5e-8)
# ran_old_seq_cojo = 
#   list.files(path='output/200910_RAN/seq_cojo/',pattern="200910_RAN.old_age") %>% 
#   purrr::map(
#   ~read_tsv(paste0("output/200910_RAN/seq_cojo/",.x),
#            col_names = c("Chr", "SNP", "bp", "freq", "refA", "b", "se", "p", "n", "freq_geno", "bJ", "bJ_se", "pJ", "LD_r"),
#            skip = 1) %>% 
#     select(SNP, CHR = Chr, BP = bp, p, pJ)) %>% 
#     purrr::reduce(bind_rows)  %>% 
#   left_join(ran_seq_gpsm_age %>% select(chrbp, q)) %>% 
#   filter(q < 0.01)

ran_young_seq_cojo = 
  list.files(path='output/200910_RAN/seq_cojo/',pattern="200910_RAN.young_age.chr") %>% 
  purrr::map(
  ~read_tsv(paste0("output/200910_RAN/seq_cojo/",.x),
           col_names = c("Chr", "SNP", "bp", "freq", "refA", "b", "se", "p", "n", "freq_geno", "bJ", "bJ_se", "pJ", "LD_r"),
           skip = 1) %>% 
    mutate(chrbp = paste(Chr, bp, sep = ":")) %>% 
    select(SNP, CHR = Chr, BP = bp, p, pJ, chrbp)) %>% 
    purrr::reduce(bind_rows) %>%  
  filter(p < 5e-8)
```

### Manhattan Plots {.tabset}

#### Log-Transformed Age Manhattan Plot

```{r}
plot_grid(
ran_seq_gpsm_logage %>% 
  ggmanhattan2(sig_threshold_p = 4.289e-9,
               sigsnps = ran_gpsm_sigsnps),
ran_seq_gpsm_logage %>% 
  ggmanhattan2(sig_threshold_p = 4.289e-9,
               sigsnps = ran_gpsm_sigsnps)+
  ylim(c(0,15)),
nrow = 2)+
  ggtitle("Red Angus Log-Transformed Age GPSM")
```


#### Raw Age 
```{r}
plot_grid(
  ran_seq_gpsm_age %>% 
    ggmanhattan2(inputfile = ., value = p, 
                 sig_threshold_p = 4.289e-9,
                 sigsnps = ran_age_seq_cojo$SNP)+
    geom_hline(yintercept = 7.3, color = "blue"),
  ran_seq_gpsm_age %>% 
    ggmanhattan2(inputfile = ., value = p, 
                 sig_threshold_p = 4.289e-9,
                 sigsnps = ran_age_seq_cojo$SNP)+
    geom_hline(yintercept = 7.3, color = "blue")+
    ylim(c(0,15)),
  nrow = 2)+
    ggtitle("Red Angus Raw Age GPSM")
```

#### Young Age 
```{r}
plot_grid(
  ran_seq_gpsm_young_age %>% 
    ggmanhattan2(inputfile = ., value = p, 
                 sig_threshold_p = 4.289e-9,
                 sigsnps = ran_young_seq_cojo$SNP)+
    geom_hline(yintercept = 7.3, color = "blue"),
  ran_seq_gpsm_young_age %>% 
    ggmanhattan2(inputfile = ., value = p, 
                 sig_threshold_p = 4.289e-9,
                 sigsnps = ran_young_seq_cojo$SNP)+
    geom_hline(yintercept = 7.3, color = "blue")+
    ylim(c(0,15)),
  nrow = 2)+
  ggtitle("Red Angus Young Animals (Post-2011) GPSM")
```

#### Old Age 
```{r}

ran_seq_gpsm_old_age %>% 
  ggmanhattan2(sig_threshold_p = 4.289e-9,
               sigsnps = ran_gpsm_sigsnps)+
  geom_hline(yintercept = 5, color = "blue")+
  ggtitle("Red Angus Old Animals (Pre-2012) GPSM")

```
#### Single Chromosome

```{r}
single_chrom = read_gwas("output/200910_RAN/seq_gwas/200910_RAN.age.chr6.mlma.gz") %>% 
  mutate(COJO = case_when(SNP %in% ran_age_seq_cojo$SNP ~ TRUE,
                          TRUE ~ FALSE),
         BP = BP/1000000)


plot_grid(
    single_chrom %>% 
      # filter(BP > 8e7,
      #        BP <10e7) %>% 
    ggmanhattan2(prune = 0.1,
                 sigsnps = single_chrom %>% 
                   filter(COJO == TRUE) %>% 
                   .$SNP,
                 sig_threshold_p = 5e-8)+
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10))+
  labs(x = "Megabases"),
  ran_nsl %>% 
    filter(CHR == 6) %>% 
    selscan_manhattans(col = abs(norm_nsl))+
    geom_vline(xintercept = single_chrom %>% 
                   filter(COJO == TRUE) %>% 
                   .$BP, 
               color = "springgreen3",
               alpha = 0.5)+
    labs(y = "Normalized nSL", x = "Megabases")+
    geom_hline(yintercept = 3, color = "red")+
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)),
  ran_xpnsl %>% 
    mutate(BP = BP/1000000) %>% 
    filter(CHR == 6,
           abs(norm_xpnsl) > 1) %>% 
    selscan_manhattans(col = abs(norm_xpnsl))+
    geom_vline(xintercept = single_chrom %>% 
                   filter(COJO == TRUE) %>% 
                   .$BP, 
               color = "springgreen3",
               alpha = 0.5)+
    labs(y = "RAiSD Mu", x = "Megabases")+
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)),
  ran_raisd %>% 
    mutate(BP = BP/1000000) %>% 
    filter(CHR == 6,
           Mu > 50) %>% 
    selscan_manhattans(col = Mu)+
    geom_vline(xintercept = single_chrom %>% 
                   filter(COJO == TRUE) %>% 
                   .$BP, 
               color = "springgreen3",
               alpha = 0.5)+
    labs(y = "RAiSD Mu", x = "Megabases")+
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)),
  nrow = 3)
```


##nSL
```{r}
ran_nsl =
  read_tsv("output/200910_RAN/selscan/nsl/200910_RAN.all.nsl.out.100bins.norm.gz",
             col_names = c("SNP", "BP", "freq", "sl1", "sl0", "nsl", "norm_nsl", "sig")) %>% 
    mutate(CHR = as.numeric(str_split_fixed(SNP, ":", n = 4)[,1]),
           SNP = paste(CHR, BP, sep = ":"),
         sig = case_when(abs(norm_nsl) > 3 ~ TRUE,
                         TRUE ~ FALSE),
         BP = BP/1000000)

ran_nsl_window = 
  read_tsv("output/200910_RAN/selscan/nsl/200910_RAN.all.nsl.out.100bins.norm.50kb.windows.gz",
           col_names = c("start", "end", "")) %>% 
  mutate(nsl_win = abs(Wstat),
         SNP = paste(CHR, BP, sep = ":")) 

```




###Genic region enrichment
I'm guessing that these are going to end up all not being significant.

```{r}
genes %>% 
  filter(chr == 1) %>% 
  mutate(bp = map2(.x = start_pos, .y = end_pos, .f = ~seq(.x,.y))) %>% 
  unnest(bp) %>%
  select(chr, bp) %>% 
  unique()
  

#Base Pairs of genes in autosomes

957545734

#Total autosomal base pairs
phyper(76, 957545734, 1619298914, 248, lower.tail = FALSE)

```


### COJO SNP overlap
```{r}
intersect(ran_age_seq_cojo$SNP, ran_young_seq_cojo$SNP) %>% 
  length()

length(ran_age_seq_cojo$SNP)
length(ran_young_seq_cojo$SNP)


length(intersect(filter(ran_seq_gpsm_age, p < 5e-8) %>% .$SNP, filter(ran_seq_gpsm_young_age, p < 5e-8) %>% .$SNP))

length(filter(ran_seq_gpsm_age, p < 5e-8) %>% .$SNP)
length(filter(ran_seq_gpsm_young_age, p < 5e-8) %>% .$SNP)
```

### Gene overlap with nSL regions

```{r}
#intersect(ran_age_seq_cojo_genes, unique(ran_raisd_genes$gene_name))

#intersect(ran_young_seq_cojo_genes, unique(ran_raisd_genes$gene_name))

map2(.x = ran_young_seq_cojo$CHR, .y = ran_young_seq_cojo$BP,
  ~filter(
    ran_nsl_sig_windows,
    CHR == .x,
    WindowStart-50000 <= .y,
    WindowStop+50000 >= .y)) %>% 
  reduce(bind_rows) 

map2(.x = ran_age_seq_cojo$CHR, .y = ran_age_seq_cojo$BP,
  ~filter(
    ran_nsl_sig_windows,
    CHR == .x,
    WindowStart-50000 <= .y,
    WindowStop+50000 >= .y)) %>% 
  reduce(bind_rows)

map2(.x = ran_young_seq_cojo$CHR, .y = ran_young_seq_cojo$BP,
  ~filter(
    ran_raisd,
    Mu > 354,
    CHR == .x,
    start <= .y,
    finish >= .y) %>% 
    mutate(COJO_SNP = paste(.x, .y, sep = ":"))) %>% 
  reduce(bind_rows) 


map2(.x = ran_age_seq_cojo$CHR, .y = ran_age_seq_cojo$BP,
  ~filter(
    ran_raisd,
    Mu > 354,
    CHR == .x,
    start <= .y,
    finish >= .y)) %>% 
  reduce(bind_rows) 


ran_raisd %>% 
  summarize(mean(finish - stat), sd(finish - stat))
```


### FAETH Score Annotations for COJO Hits
1) Write out SNP names, will just use grep in `joinable_FAETH_ARS.txt` file to find these SNPs as reading in isn't exactly helpful
2) Read them in and compare to distribution of FAETH scores for each 
Commands:

```{bash}
> grep  "1:720042 \|1:148638095 \|10:509620 \|10:3304349 \|10:89215226 \|11:17539086 \|11:59119877 \|13:5262550 \|13:11822579 \|13:36339812 \|13:37848629 \|14:22518 \|14:156254 \|14:4050892 \|14:11886077 \|14:23257180 \|16:21295945 \|16:68631334 \|18:19529870 \|19:412779 \|2:6757167 \|2:7498211 \|2:8100524 \|2:21860593 \|2:111403008 \|2:113671290 \|2:127446835 \|20:30112844 \|20:65179823 \|21:1148483 \|21:2117530 \|21:6722575 \|21:7103453 \|21:14552611 \|21:39317383 \|21:62593095 \|23:1032665 \|23:1366335 \|23:1366691 \|23:1768070 \|23:1783688 \|23:1805412 \|23:48398507 \|25:8955090 \|25:9065380 \|26:32335380 \|26:32357492 \|26:33408375 \|26:34425381 \|27:405898 \|28:31307728 \|28:42523583 \|3:91361195 \|3:112272306 \|4:4055714 \|5:85176900 \|6:11410294 \|6:12842123 \|6:25861907 \|6:56961317 \|6:59577310 \|6:114937887 \|7:26806870 \|7:56681778 \|7:88102065 \|7:88897740 \|7:89137291 \|8:69385 \|9:846219 \|9:15850396 \|9:39591442 \|9:95146355" data/FAETH_intermediates/joinable_FAETH_ARS.txt.gz > output/200910_RAN/seq_cojo/FAETH/200910_RAN.age.cojo.FAETH.txt

> zgrep "1:722538 \|10:3316433 \|10:5423142 \|10:7702688 \|10:16404552 \|10:89215226 \|11:17592781 \|11:17592868 \|11:17849450 \|11:20569541 \|11:69399750 \|12:442568 \|12:46730507 \|13:1113875 \|13:11822579 \|13:37848629 \|13:46808594 \|14:22518 \|14:156254 \|14:8506714 \|14:23257180 \|14:59747342 \|14:59788659 \|14:78037164 \|14:78044125 \|15:47481607 \|15:64765703 \|15:64781739 \|16:957211 \|16:73942705 \|17:51861826 \|18:63778412 \|19:242064 \|19:412779 \|2:7500278 \|2:8100524 \|2:22339424 \|2:53126236 \|2:53151542 \|2:53189373 \|2:53283910 \|2:53803504 \|2:103217345 \|2:103225507 \|2:136159490 \|20:49419382 \|21:1146167 \|21:2117530 \|21:3887410 \|21:64875723 \|21:64957035 \|23:490909 \|23:838197 \|23:1324051 \|23:1366335 \|23:1366691 \|23:1768070 \|23:1800642 \|23:1832160 \|23:2082004 \|23:2599415 \|24:27780789 \|24:38012259 \|24:38947749 \|24:38971013 \|24:39021058 \|25:8738478 \|25:9065380 \|25:36599479 \|26:32346219 \|26:32357492 \|26:34205775 \|26:34735344 \|26:45137144 \|26:49325027 \|28:613012 \|3:7808113 \|3:13344300 \|3:64252547 \|3:91361195 \|3:91481076 \|3:117849750 \|4:50134171 \|5:53033839 \|6:842877 \|6:49826827 \|6:49885440 \|6:89631901 \|6:114921606 \|7:358667 \|7:369503 \|7:88800495 \|9:15850396 \|9:23244267 \|9:39413842 \|9:82349415" data/FAETH_intermediates/joinable_FAETH_ARS.txt.gz > output/200910_RAN/seq_cojo/FAETH/200910_RAN.young_age.cojo.FAETH.txt
```

The top 1/2 FAETH score cutoff = 1.48e-8
The top 1/3 FAETH score cutoff = 1.60e-8
```{r}
ran_age_seq_cojo_FAETH = 
  read_delim("output/200910_RAN/seq_cojo/FAETH/200910_RAN.age.gw_cojo.FAETH.txt",
           col_names = c("SNP", "UMDSNP", "CHR", "BP", "bull.FAETH1.1", "cow.FAETH1.1", "annotation", "aseQTL", "eeQTL", "geQTL", "bull.ldscore50kb", "bull.maf", "ChIPSeq", "mQTL", "selection.sig", "young", "variant.density50kb", "sQTL", "conserved100way", "cow.ldscore50kb", "cow.maf"), 
           col_type = cols(SNP = col_character()),
           delim = " ") %>% 
  mutate(FAETH = (bull.FAETH1.1 + cow.FAETH1.1)/2,
         top_half = case_when(FAETH >= 1.48e-8 ~ TRUE,
                              TRUE ~ FALSE),
         top_third = case_when(FAETH >= 1.60e-8 ~ TRUE,
                              TRUE ~ FALSE))

ran_young_seq_cojo_FAETH = 
  read_delim("output/200910_RAN/seq_cojo/FAETH/200910_RAN.young_age.gw_cojo.FAETH.txt",
           col_names = c("SNP", "UMDSNP", "CHR", "BP", "bull.FAETH1.1", "cow.FAETH1.1", "annotation", "aseQTL", "eeQTL", "geQTL", "bull.ldscore50kb", "bull.maf", "ChIPSeq", "mQTL", "selection.sig", "young", "variant.density50kb", "sQTL", "conserved100way", "cow.ldscore50kb", "cow.maf"),
           col_type = cols(SNP = col_character()),
           delim = " ") %>% 
  mutate(FAETH = (bull.FAETH1.1 + cow.FAETH1.1)/2,
         top_half = case_when(FAETH >= 1.48e-8 ~ TRUE,
                              TRUE ~ FALSE),
         top_third = case_when(FAETH >= 1.60e-8 ~ TRUE,
                              TRUE ~ FALSE))


ran_age_seq_cojo_FAETH %>% count(top_half)
ran_age_seq_cojo_FAETH %>% count(top_third)
ran_young_seq_cojo_FAETH %>% count(top_half)
ran_young_seq_cojo_FAETH %>% count(top_third)


binom.test()
```

```{r}
ran_age_seq_cojo_FAETH %>% 
  left_join(ran_age_seq_cojo) %>% 
  View()
```


### Gene Annotations for COJO hits:
 
For outputting unique gene lists
```{r} 
ran_age_seq_cojo %>% 
  gene_annotation(search_bp = 50000) %>%
  #   left_join(ran_age_seq_cojo, 
  #             by = c("CHR", "BP")) %>% 
  # select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% 
  select(gene_name) %>% 
  # filter(!is.na(gene_name)) %>% 
  unique() %>% 
  write_tsv("output/200910_RAN/genes/200910_RAN.age.gpsm.genes.50kb.txt",
            col_names = FALSE)

ran_young_seq_cojo %>% 
  gene_annotation(search_bp = 50000) %>%
    left_join(ran_age_seq_cojo, 
              by = c("CHR", "BP")) %>% 
  select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% 
  select(gene_id) %>% 
  filter(!is.na(gene_name)) %>% 
  unique() %>% 
  write_tsv("output/200910_RAN/genes/200910_RAN.young_age.gpsm.genes.50kb.txt",
            col_names = FALSE)

#Output for genewalk
read_tsv("output/200910_RAN/genes/200910_RAN.age.gpsm.genes.50kb.txt") %>% 
  arrange(log10(p)) %>% 
  filter(!is.na(gene_name)) %>% 
  select(gene_id) %>% .$gene_id %>%  
  gprofiler2::gorth(source_organism = "btaurus", target_organism = "hsapiens") %>% 
  select(ortholog_ensg) %>% 
  unique() %>% 
  write_tsv("output/200910_RAN/genes/200910_RAN.age.gpsm.genes.50kb.humanortho.txt", col_names = FALSE)

 read_tsv("output/200910_RAN/genes/200910_RAN.age.gpsm.cojogenes.50kb.txt") %>% 
  arrange(log10(p)) %>% 
  filter(!is.na(gene_name)) %>% 
  select(gene_id) %>% .$gene_id %>%  
  gprofiler2::gorth(source_organism = "btaurus", target_organism = "hsapiens") %>% 
  select(ortholog_ensg) %>% 
  unique() %>% 
  write_tsv("output/200910_RAN/genes/200910_RAN.young_age.gpsm.genes.50kb.humanortho.txt", col_names = FALSE)   
    
ran_young_seq_cojo %>% 
  gene_annotation(search_bp = 50000) %>% 
    left_join(ran_young_seq_cojo, 
              by = c("CHR", "BP")) %>% 
  select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% 
  select(gene_name) %>% 
  filter(!is.na(gene_name)) %>% 
  unique() %>% 
  write_tsv("output/200910_RAN/genes/200910_RAN.young_age.gpsm.genes.50kb.txt",
            col_names = FALSE)
```


Annotated genes within 50 kb of COJO SNP

#### Age Genes
```{r}
ran_young_seq_cojo %>% 
  gene_annotation() %>% 
    right_join(ran_young_seq_cojo, 
              by = c("CHR", "BP")) %>% 
  select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% write_tsv("output/200910_RAN/genes/200910_RAN.young_age.gpsm.cojogenes.50kb.txt") 
  datatable()

ran_age_seq_cojo %>% 
  gene_annotation() %>% 
    right_join(ran_age_seq_cojo, 
              by = c("CHR", "BP")) %>%
  mutate(ingene = case_when(distance == 0 ~ "within",
                            is.na(distance) ~ ">50kb",
                            TRUE ~ "proximal")) %>% 
  group_by(SNP) %>% 
  top_n(1, wt = -distance) %>% 
  ungroup() %>% 
  count(ingene)

ran_young_seq_cojo %>% 
  gene_annotation() %>% 
    right_join(ran_young_seq_cojo, 
              by = c("CHR", "BP")) %>%
  mutate(ingene = case_when(distance == 0 ~ "within",
                            is.na(distance) ~ ">50kb",
                            TRUE ~ "proximal")) %>% 
  group_by(SNP) %>% 
  top_n(1, wt = -distance) %>% 
  ungroup() %>% 
  count(ingene)

```
### Shared domains with selective sweep meta-analysis

#### Age
```{r}
ran_age_seq_cojo_genes =
  ran_age_seq_cojo %>% 
    gene_annotation() %>% 
      left_join(ran_age_seq_cojo, 
                by = c("CHR", "BP")) `%>% 
    select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% 
  select(gene_name) %>% 
  distinct() %>% 
  filter(!is.na(gene_name)) %>% 
  .$gene_name

sweeps = read_csv("data/meata_sweeeps.csv")

ran_age_seq_cojo_genes %>% 
  purrr::map(
    ~filter(sweeps,str_detect(`Gene Names`, .x)) %>% 
      mutate(driver = .x)
    ) %>% 
  reduce(bind_rows) %>% 
  count(driver)
   
    
```

#### Young Age
```{r}
ran_young_seq_cojo_genes =
  ran_young_seq_cojo %>% 
    gene_annotation() %>% 
      left_join(ran_young_seq_cojo, 
                by = c("CHR", "BP")) %>% 
    select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% 
  select(gene_name) %>% 
  distinct() %>%
  filter(!is.na(gene_name)) %>% 
  .$gene_name

sweeps = read_csv("data/meata_sweeeps.csv")
  
small_sweeps = read_csv("data/meata_sweeeps.csv") %>% 
  mutate(size = as.numeric(str_split_fixed(`Validated MSS region (Mb)`, pattern = "-", n = 2)[,2]) - as.numeric(str_split_fixed(`Validated MSS region (Mb)`, pattern = "-", n = 2)[,1])) %>% 
  filter(size < 2.5)

ran_young_seq_cojo_genes %>% 
  purrr::map(
    ~filter(sweeps,str_detect(`Gene Names`, .x)) %>% 
      mutate(driver = .x)
    ) %>% 
  reduce(bind_rows) %>% 
  count(driver)

ran_young_seq_cojo_genes %>% 
  purrr::map(
    ~filter(small_sweeps,str_detect(`Gene Names`, .x)) %>% 
      mutate(driver = .x)
    ) %>% 
  reduce(bind_rows) %>% 
  count(driver)
   
sweeps %>% 
  ggplot(aes(size))+
  geom_histogram()
```

### Shared Genes beetween FULL and YOUNG datasets
```{r}
intersect(ran_age_seq_cojo_genes, ran_young_seq_cojo_genes)

length(ran_young_seq_cojo_genes)
length(ran_age_seq_cojo_genes)

bind_rows(ran_age_seq_cojo %>% 
            filter(p < 5e-8) %>% 
            mutate(dataset = "AGE"), 
          ran_young_seq_cojo %>% 
            filter(p < 5e-8) %>% 
            mutate(dataset = "YOUNG")) %>% 
  arrange(CHR,BP) %>% 
  group_by(CHR) %>% 
  mutate(distance = BP - lag(BP, 1)) %>% View()
  filter(dataset == "AGE",
         distance > 1000000)
```


### Gene Ontology Analysis

Gene ontology enrichment analysis using `gprofiler2::gost()` of gene names from GALLO annotation of log-transformed GPSM COJO hits

#### Full Age
Significance threshold set at FDR < 0.10
```{r, eval }
ran_age_seq_cojo_GO =
  ran_age_seq_cojo %>% 
  gene_annotation(search_bp = 50000) %>% 
    left_join(ran_age_seq_cojo, 
              by = c("CHR", "BP")) %>% 
  select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% 
  .$gene_name %>% unique() %>% 
  gost(query = ., 
      organism = "btaurus",
      user_threshold = 0.1,
      correction_method = "fdr")


gostplot(ran_age_seq_cojo_GO, capped = TRUE, interactive = TRUE)
```

#### Young Age
Significance threshold set at FDR < 0.10
```{r, eval }
ran_young_seq_cojo_GO =
  ran_young_seq_cojo %>% 
  gene_annotation(search_bp = 50000) %>% 
    left_join(ran_young_seq_cojo, 
              by = c("CHR", "BP")) %>% 
  select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% 
  .$gene_name %>% unique() %>% 
  gost(query = ., 
      organism = "btaurus",
      user_threshold = 0.1,
      correction_method = "fdr")


gostplot(ran_young_seq_cojo_GO, capped = TRUE, interactive = TRUE)
```

### QTL Enrichments {.tabset}

Results are for Red Angus sequence-based COJO hits with p < 1e-5. QTL are within 50 kb of identified SNP

Removed milk-associated QTL, perform enrichment on whole genome QTL and by chromosome

#### Age (Genome)
```{r}
ran_age_seq_cojo %>%
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>%
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  select(QTL, adj.pval, pvalue, N_QTLs, N_QTLs_db, QTL_type)

ran_age_seq_cojo %>%
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>%
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "chromosome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  select(QTL, CHR, adj.pval, pvalue, N_QTLs, N_QTLs_db, QTL_type) %>% 
  group_by(QTL) %>% 
  top_n(1, wt = N_QTLs) %>% 
  arrange(adj.pval) %>% 
  View()
```

#### Young (Genome)
```{r}
ran_young_seq_cojo %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  select(QTL, adj.pval, pvalue, N_QTLs, N_QTLs_db, QTL_type)

ran_young_seq_cojo %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "chromosome", nThreads = 2)%>% 
  arrange(adj.pval) %>% 
  select(QTL, CHR, adj.pval, pvalue, N_QTLs, N_QTLs_db, QTL_type) %>% 
  group_by(QTL) %>% 
  top_n(1, wt = N_QTLs) %>% 
  arrange(adj.pval) %>% 
  View()
```

#### Age (per-Chromosome)
```{r}
ran_age_seq_cojo %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "chromosome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  select(QTL, CHR, adj.pval, pvalue, N_QTLs, N_QTLs_db, QTL_type)

ran_age_seq_cojo %>% 
  qtl_annotation() %>% 
  filter(trait_ID == "Calving ease")
```

#### Young (per-chromosme)
```{r}
qtls = 
ran_young_seq_cojo %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") 


qtl_enrich =
  qtls %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2)
  
qtl_enrich %>% 
  filter(adj.pval < 0.05) %>% 
  arrange(adj.pval) %>% 
  left_join(qtls %>% 
              select(CHR, BP, QTL = trait_ID, QTL_ID)) %>% 
  group_by(QTL, N_QTLs, adj.pval) %>% 
  summarize(n(), n_distinct(QTL_ID))
  
  
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "chromosome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()

```

### Annnotating with FAANG data

```{r}
ran_age_seq_cojo_functional = 
  map2(.x = ran_age_seq_cojo$CHR, .y = ran_age_seq_cojo$BP,
    ~filter(
      FAANG,
      CHR == .x,
      START <= .y,
      STOP >= .y)%>% 
      mutate(COJO_SNP = paste(.x, .y, sep = ":"))) %>% 
    reduce(bind_rows) %>% 
  left_join(ran_age_seq_cojo,
            by = c("COJO_SNP" = "SNP")) 


ran_young_seq_cojo_functional = 
  map2(.x = ran_young_seq_cojo$CHR, .y = ran_young_seq_cojo$BP,
    ~filter(
      FAANG,
      CHR == .x,
      START <= .y,
      STOP >= .y)%>% 
      mutate(COJO_SNP = paste(.x, .y, sep = ":"))) %>% 
    reduce(bind_rows) %>% 
  left_join(ran_young_seq_cojo, 
            by = c("COJO_SNP" = "SNP"))



ran_age_seq_cojo_functional %>% 
  select(COJO_SNP,TISSUE) %>% 
  distinct() %>% 
  mutate(case = as.integer(1)) %>% 
  pivot_wider(names_from = TISSUE, values_from = case, values_fill = as.integer(0)) %>% 
  as.data.frame() %>% 
  upset(nsets = 8, order.by = "freq")
  
  
ran_young_seq_cojo_functional %>% 
  select(COJO_SNP,TISSUE) %>% 
  distinct() %>% 
  mutate(case = as.integer(1)) %>% 
  pivot_wider(names_from = TISSUE, values_from = case, values_fill = as.integer(0)) %>% 
  as.data.frame() %>% 
  upset(nsets = 8, order.by = "freq")

ran_young_seq_cojo_functional %>% 
  select(COJO_SNP,MARK) %>% 
  distinct() %>% 
  mutate(case = as.integer(1)) %>% 
  pivot_wider(names_from = MARK, values_from = case, values_fill = as.integer(0)) %>% 
  as.data.frame() %>% 
  upset(nsets = 8, order.by = "freq")


ran_young_seq_cojo_functional %>%
  mutate(MARK = paste(TISSUE, MARK, sep = ":")) %>% 
  select(COJO_SNP,MARK) %>% 
  distinct() %>% 
  mutate(case = as.integer(1)) %>% 
  pivot_wider(names_from = MARK, values_from = case, values_fill = as.integer(0)) %>% 
  as.data.frame() %>% 
  upset(nsets = 20, order.by = "freq")

```
### Hypergeometric test of significant marks

Base pairs covered by each of the following epigenetic marks:
H3K27ac = 225,723,813 bp, ~8% of genome, binomial p-value = 6.6e-7
H3K4me1 = 182,645,598

```{r}
# ATAC
#binom.test(34, 248, , alternative = "greater")
# CTCF
binom.test(18, 248, 0.067, alternative = "greater")
# H3K27ac
binom.test(16, 248, 0.081, alternative = "greater")
# H3K27me3
#binom.test(63, 248, , alternative = "greater")
# H3K4me1
binom.test(17, 248, 0.065, alternative = "greater")
# H3K4me3
binom.test(5, 248, 0.026, alternative = "greater")
```
#### For young animals only:
```{r}
# ATAC
#binom.test(84, 417, , alternative = "greater")
# CTCF
binom.test(31, 417, 0.067, alternative = "greater")
# H3K27ac
binom.test(32, 417, 0.081, alternative = "greater")
# H3K27me3
#binom.test(112, 417, , alternative = "greater")
# H3K4me1
binom.test(28, 417, 0.065, alternative = "greater")
# H3K4me3
binom.test(15, 417, 0.026, alternative = "greater")
```


### Testing significance of annotated marks
```{r}
seq_positions = read_tsv("data/200910_RAN/200910_RAN.young_age.cojosnps.bim",
                         col_names = c("CHR", "SNP", "CM","BP", "REF", "ALT")) %>% 
  select(CHR, BP)
test_positions =
  seq_positions %>% 
  sample_n(10)

faang_df =
    map2(.x = test_positions$CHR, .y = test_positions$BP,
    ~filter(
      FAANG,
      CHR == .x,
      START <= .y,
      STOP >= .y) %>% 
      mutate(SNP = paste(.x, .y, sep = ":"))) %>% 
    reduce(bind_rows)

epi_tester = function(rep){
  test_positions =
    seq_positions %>% 
    sample_n(417)
  faang_df =
    map2(.x = test_positions$CHR, .y = test_positions$BP,
    ~filter(
      FAANG,
      CHR == .x,
      START <= .y,
      STOP >= .y) %>% 
      mutate(SNP = paste(.x, .y, sep = ":"))) %>% 
    reduce(bind_rows)
  
  return(data.frame(rep = rep, npeaks = count(faang_df), nsnps = faang_df %>% count(SNP) %>% count()))
}

faang_tests = 
  seq(1,100) %>% 
    purrr::map(~epi_tester(.x)) %>% 
    reduce(bind_rows)

```


## Conditional Joint Analysis (COJO)


### Candidate genes associated with COJO hits:

Using GALLO annotation to identify nearby genes (within 50 kb) to significant (p < 1e-5) COJO hits

#### Log Age
```{r}

ran_gpsm_log_age_cojo = 
  read_tsv("output/200910_RAN/cojo/200910_RAN.log_age.850K.jma.cojo.gz") %>% 
    select(CHR = Chr, BP = bp, p = pJ) %>% 
    gene_annotation() %>% 
    left_join(read_tsv("output/200910_RAN/cojo/200910_RAN.log_age.850K.jma.cojo.gz"), 
              by = c("CHR" = "Chr", "BP" = "bp")) %>% 
  select(-chr, -start_pos, -end_pos, -SNP, -gene_biotype, -width, -strand, -upstream, -downstream, -refA, -b, -se, -freq_geno, -n)
dim(ran_gpsm_log_age_cojo)

datatable(ran_gpsm_age_cojo)
```

#### Raw Age
```{r}

ran_gpsm_age_cojo = 
  read_tsv("output/200910_RAN/cojo/200910_RAN.age.850K.jma.cojo.gz") %>% 
    select(CHR = Chr, BP = bp, p = pJ) %>% 
    gene_annotation() %>% 
    left_join(read_tsv("output/200910_RAN/cojo/200910_RAN.age.850K.jma.cojo.gz"), 
              by = c("CHR" = "Chr", "BP" = "bp")) %>% 
  select(-chr, -start_pos, -end_pos, -SNP, -gene_biotype, -width, -strand, -upstream, -downstream, -refA, -b, -se, -freq_geno, -n)
dim(ran_gpsm_age_cojo)

datatable(ran_gpsm_age_cojo)
```


### Gene Ontology Analysis

Gene ontology enrichment analysis using `gprofiler2::gost()` of gene names from GALLO annotation of log-transformed GPSM COJO hits

Significance threshold set at FDR < 0.10
```{r}
GO_results = gost(query = ran_gpsm_age_cojo$gene_name, 
                  organism = "btaurus",
                  user_threshold = 0.1,
                  correction_method = "fdr")


gostplot(GO_results, capped = TRUE, interactive = TRUE)
```

### QTL-enrichment analysis (within annotated QTL)

These QTL are quite small (only 4 bp), things certainly change when we search for a larger area around them.
```{r}
read_tsv("output/200910_RAN/cojo/200910_RAN.age.850K.jma.cojo.gz") %>% 
  select(CHR = Chr, BP = bp, p = pJ) %>% 
  qtl_annotation(search_bp = 0) %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```
### QTL-enrichment analysis (within 50kb annotated QTL)

These QTL are quite small (only 4 bp) we identify much many more when we're searching in a substantial area around them.
```{r}
read_tsv("output/200910_RAN/cojo/200910_RAN.age.850K.jma.cojo.gz") %>% 
  select(CHR = Chr, BP = bp, p = pJ) %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```

### Per-Chromosome QTL-enrichment analysis (within 50kb of significant SNP)
```{r}
read_tsv("output/200910_RAN/cojo/200910_RAN.age.850K.jma.cojo.gz") %>% 
  select(CHR = Chr, BP = bp, p = pJ) %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "chromosome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```

### Tajima's D

```{r}
RAN_Tajima = read_tsv("output/200910_RAN/200910_RAN.1kbulls.Tajima.D")
filter(RAN_Tajima, 
       CHROM == 23, BIN_START < 2000000, BIN_START > 1000000) %>% 
  filter(TajimaD > 2)
CHR23TD %>% 
  select(CHR = CHROM, BP = BIN_START, TajimaD, SNP = N_SNPS) %>%
  selscan_manhattans(col = TajimaD)

RAN_Tajima %>% 
  ggplot(aes(TajimaD))+
  geom_density()
RAN_Tajima %>% 
  top_frac(0.1, wt = TajimaD) %>% 
  .$TajimaD %>% 
  min()

```


