---
title: "GPSM"
author: "Troy Rowan"
date: "2020-09-20"
output: 
  html_document:
    code_folding: show
editor_options:
  chunk_output_type: inline
  markdown: 
    wrap: 72
---

```{r, setup, include = FALSE}
set.seed(325333)
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
knitr::opts_knit$set(root.dir = here::here())
library(workflowr)
library(knitr)
library(maps)
library(ggthemes)
library(purrr)
library(factoextra)
library(corrr)
library(ggcorrplot)
library(factoextra)
library(here)
library(tidyverse)
library(lubridate)
library(pedigree)
library(qvalue)
library(GALLO)
library(cowplot)
library(gprofiler2)
library(DT)
library(ggforce)
library(vroom)
#source("../code/annotation_functions.R")
```

```{r}
source("code/GCTA_functions.R")
source("code/annotation_functions.R")
simmental = read_csv("output/200907_SIM/phenotypes/200907_SIM.info.csv")
redangus = read_csv("output/200910_RAN/phenotypes/200910_RAN.info.csv")
```
# Red Angus GPSM Analysis

## REML variance component estimates {.tabset}

### Raw Age

Raw Age (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 5.244    | 0.128 |
| V(e)    | 4.789    | 0.040 |
| V(p)    | 10.03    | 0.118 |
| V(G)/Vp | 0.523    | 0.007 |

### Square Root

Square Root Transformed Age (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.242    | 0.005 |
| V(e)    | 0.151    | 0.001 |
| V(p)    | 0.393    | 0.005 |
| V(G)/Vp | 0.616    | 0.006 |

### Cube Root

Cube Root Transformed Age (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.065    | 0.001 |
| V(e)    | 0.037    | 0.000 |
| V(p)    | 0.101    | 0.001 |
| V(G)/Vp | 0.637    | 0.006 |

### Box-Cox

Box-Cox Transformed Age (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.115    | 0.002 |
| V(e)    | 0.060    | 0.001 |
| V(p)    | 0.175    | 0.002 |
| V(G)/Vp | 0.657    | 0.006 |

### Log

Log Transformed Age (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.222    | 0.005 |
| V(e)    | 0.115    | 0.001 |
| V(p)    | 0.337    | 0.005 |
| V(G)/Vp | 0.657    | 0.006 |

### Young Raw Age 

Raw Age (Born 2012-Present) (n= 44,470):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 3.354    | 0.061 |
| V(e)    | 0.983    | 0.001 |
| V(p)    | 4.337    | 0.057 |
| V(G)/Vp | 0.773    | 0.004 |

### Young Log Age

Log Transformed Age (Born 2012-Present) (n= 46,454):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.200    | 0.004 |
| V(e)    | 0.079    | 0.001 |
| V(p)    | 0.279    | 0.004 |
| V(G)/Vp | 0.718    | 0.005 |

### Old Raw Age 

Raw Age (Born 2012-Present) (n= 1,984):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 16.70    | 1.277 |
| V(e)    | 13.30    | 0.854 |
| V(p)    | 30.00    | 1.041 |
| V(G)/Vp | 0.557    | 0.031 |

### Old Log Age

Log Transformed Age (Born 2012-Present) (n= 1,984):

| Source  | Variance | SE    |
|---------|----------|-------|
| V(G)    | 0.056    | 0.004 |
| V(e)    | 0.045    | 0.003 |
| V(p)    | 0.100    | 0.003 |
| V(G)/Vp | 0.555    | 0.031 |

## Individual Residuals and Breeding Values {.tabset}

These are REML estimates of individual's breeding values and residuals
from GCTA GREML analysis

### Age

```{r}
plot_grid(
  read_blp("output/200910_RAN/greml/200910_RAN.age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Raw Age GPSM\nResiduals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nBreeding Values")+
    theme_cowplot())
```

### Square Root

```{r}
plot_grid(
  read_blp("output/200910_RAN/greml/200910_RAN.sqrt_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Square Root Transformed Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.sqrt_age.850K.indi.blp") %>% 
  left_join(redangus %>% 
              select(international_id, age)) %>% 
  ggplot(aes(sample = BV))+
  stat_qq()+
  stat_qq_line(color = "red")+
  ggtitle("\nBreeding Values")+
  theme_cowplot())
```

### Cube Root

```{r}
plot_grid(
  read_blp("output/200910_RAN/greml/200910_RAN.cbrt_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Cube Root Transformed Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.cbrt_age.850K.indi.blp") %>% 
  left_join(redangus %>% 
              select(international_id, age)) %>% 
  ggplot(aes(sample = BV))+
  stat_qq()+
  stat_qq_line(color = "red")+
  ggtitle("\nGPSM Breeding Values")+
  theme_cowplot())
```

### Box-Cox

```{r}

plot_grid(
  read_blp("output/200910_RAN/greml/200910_RAN.bc_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Box-Cox Transformed Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.bc_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Residuals")+
    theme_cowplot())
```

### Log

```{r}

plot_grid(read_blp("output/200910_RAN/greml/200910_RAN.log_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Log Transformed Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.log_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```

### Youg Raw Age

```{r}

plot_grid(read_blp("output/200910_RAN/greml/200910_RAN.young_age.850K.indi.blp.gz") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("(2012-Present) Animals Only\n Raw Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.young_age.850K.indi.blp.gz") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```


### Log

```{r}

plot_grid(read_blp("output/200910_RAN/greml/200910_RAN.young_log_age.850K.indi.blp.gz") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("Log Transformed Age \nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.young_log_age.850K.indi.blp.gz") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```


### Old Raw Age

```{r}

plot_grid(read_blp("output/200910_RAN/greml/200910_RAN.old_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("(2011 and Earlier)\n Raw Age\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.old_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())
```

### Old log Age

```{r}

plot_grid(read_blp("output/200910_RAN/greml/200910_RAN.old_log_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = Residual))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("(2011 and Earlier)\n Raw Age\nGPSM Residuals")+
    theme_cowplot(),
  read_blp("output/200910_RAN/greml/200910_RAN.old_log_age.850K.indi.blp") %>% 
    left_join(redangus %>% 
                select(international_id, age)) %>% 
    ggplot(aes(sample = BV))+
    stat_qq()+
    stat_qq_line(color = "red")+
    ggtitle("\nGPSM Breeding Values")+
    theme_cowplot())


```


## Breeding Value Correlations Across Transforms/SUbsets  

```{r}
blup_compare =
  read_blp("output/200910_RAN/greml/200910_RAN.young_age.850K.indi.blp.gz") %>% 
    left_join(read_blp("output/200910_RAN/greml/200910_RAN.log_age.850K.indi.blp") %>% 
                select(international_id, logageBV = BV, logageResidual = Residual))

cor(blup_compare$BV, blup_compare$logageBV)

old_blups =
  read_blp("output/200910_RAN/greml/200910_RAN.age.850K.indi.blp") %>%
  filter(international_id %in% read_blp("output/200910_RAN/greml/200910_RAN.old_age.850K.indi.blp")$international_id) %>% 
    left_join(read_blp("output/200910_RAN/greml/200910_RAN.log_age.850K.indi.blp") %>% 
                select(international_id, logageBV = BV, logageResidual = Residual))

cor(old_blups$BV, old_blups$logageBV)

young_blups =
  read_blp("output/200910_RAN/greml/200910_RAN.age.850K.indi.blp") %>%
  filter(!international_id %in% read_blp("output/200910_RAN/greml/200910_RAN.old_age.850K.indi.blp")$international_id) %>% 
    left_join(read_blp("output/200910_RAN/greml/200910_RAN.log_age.850K.indi.blp") %>% 
                select(international_id, logageBV = BV, logageResidual = Residual))

cor(young_blups$BV, young_blups$logageBV)

```


## n(SigSNPs)

The number of significant SNPs in each analysis at various significance
level cutoffs for both p and q values

|        | p\<1e-5 | p\<1e7.55e-7 | q\<0.1 | q\<0.05 |
|--------|---------|--------------|--------|---------|
| Raw    | 315     | 214          | 509    | 398     |
| Sqrt   | 453     | 333          | 729    | 559     |
| Cbrt   | 471     | 357          | 817    | 596     |
| BoxCox | 540     | 390          | 907    | 754     |
| Log    | 513     | 377          | 822    | 715     |
| YAge   | 762     | 555          | 1210   | 1042    |
| YLog   | 743     | 533          | 1123   | 989     |
| OAge   | 762     | 555          | 1210   | 1042    |
| OLog   | 743     | 533          | 1123   | 989     |

```{r eval=TRUE, echo=FALSE}
sigcheck = function(df){
  gwasdf = 
    df %>% 
    mutate(p_1e5 = case_when(p <= 1e-5 ~ TRUE,
                             TRUE ~ FALSE),
           bonferr = case_when(p <= 7.55e-7 ~ TRUE,
                             TRUE ~ FALSE),
           q01 = case_when(q <= 0.10 ~ TRUE,
                             TRUE ~ FALSE),
           q005 = case_when(q <= 0.05 ~ TRUE,
                             TRUE ~ FALSE))%>% 
  select(p_1e5, bonferr, q01, q005) %>% 
  colSums() 
  return(gwasdf)
}

ran_gpsm_age = 
  read_gwas("output/200910_RAN/gwas/200910_RAN.age.850K.mlma.gz")

ran_gpsm_sqrtage = 
  read_gwas("output/200910_RAN/gwas/200910_RAN.sqrt_age.850K.mlma.gz")

ran_gpsm_cbrtage = 
  read_gwas("output/200910_RAN/gwas/200910_RAN.cbrt_age.850K.mlma.gz")

ran_gpsm_bcage = 
  read_gwas("output/200910_RAN/gwas/200910_RAN.bc_age.850K.mlma.gz")

ran_gpsm_logage = 
  read_gwas("output/200910_RAN/gwas/200910_RAN.log_age.850K.mlma.gz")

ran_gpsm_youngage = 
  read_gwas("output/200910_RAN/gwas/200910_RAN.young_age.850K.mlma.gz")

ran_gpsm_younglogage = 
  read_gwas("output/200910_RAN/gwas/200910_RAN.young_log_age.850K.mlma.gz")

ran_gpsm_oldage = 
  read_gwas("output/200910_RAN/gwas/200910_RAN.old_age.850K.mlma.gz")

ran_gpsm_oldlogage = 
  read_gwas("output/200910_RAN/gwas/200910_RAN.old_log_age.850K.mlma.gz")

rbind(sigcheck(ran_gpsm_age),
      sigcheck(ran_gpsm_sqrtage),
      sigcheck(ran_gpsm_cbrtage),
      sigcheck(ran_gpsm_bcage),
      sigcheck(ran_gpsm_logage),
      sigcheck(ran_gpsm_youngage),
      sigcheck(ran_gpsm_younglogage),
      sigcheck(ran_gpsm_oldage),
      sigcheck(ran_gpsm_oldlogage)) %>% as.data.frame() %>% 
mutate(analysis = c("ran_gpsm_age", "ran_gpsm_sqrtage", "ran_gpsm_cbrtage","ran_gpsm_bcage", "ran_gpsm_logage", "ran_gpsm_youngage", "ran_gpsm_oldage", "ran_gpsm_oldlogage")) %>% 
select(analysis, everything()) %>% 
datatable(caption = "Number of significant SNPs at various significance levels with different dependent variables")
  
```


## GPSM GWAS Manhattan Plots for Transformed Age {.tabset}

### Raw Age

(Significance threshold - Bonferroni)

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_age,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07),
  ggmanhattan2(ran_gpsm_age,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07)+
    ylim(c(0,15)),
  nrow = 2)

#Saving significant SNPs for highlighting in other plots:
raw_age_sigsnps = 
  ran_gpsm_age %>% 
    filter(p < 7.546167e-07) %>% .$SNP
```

### Square Root

Square root transformed age as phenotype (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_sqrtage,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_sqrtage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_sqrtage,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_sqrtage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)
```

### Cube Root

Cube Root Transformed Age Manahattan Plots (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_cbrtage,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_cbrtage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_cbrtage,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_cbrtage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)
```

### Box-Cox

Box-Cox Transformed Age Manahattan Plots (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_bcage,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_bcage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_bcage,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_bcage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)
```

### Log

Log Transformed Age Manahattan Plots (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_logage %>% 
                 filter(CHR <= 29),
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_logage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_logage%>% 
                 filter(CHR <= 29),
               prune = 0.1,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_logage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)
ggsave("output/figures/200910_RAN.logage.850K.manhattan.png", width = 12, height = 8)


plot_grid(
  ggmanhattan2(ran_gpsm_logage,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07),
  ggmanhattan2(ran_gpsm_logage,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07)+
    ylim(c(0,15)),
  nrow = 2)

logage_sigsnps = 
  filter(ran_gpsm_logage,
         p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% .$SNP
  
```

### Young Raw Age

Young animal only (Born after Jan 1, 2012) Raw Age Manahattan Plots (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_youngage,
               prune = 0.01, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & SNP %in% c(raw_age_sigsnps, logage_sigsnps)) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_youngage,
               prune = 0.01,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & SNP %in% c(raw_age_sigsnps, logage_sigsnps)) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)

plot_grid(
  ggmanhattan2(ran_gpsm_youngage,
               prune = 0.01, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = logage_sigsnps),
  ggmanhattan2(ran_gpsm_youngage,
               prune = 0.01,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = logage_sigsnps)+
    ylim(c(0,15)),
  nrow = 2)
```

### Old  Age

Old animal only (Born before Jan 1, 2012) Raw Age and log-transformed Manhattan Plots (Significance threshold - 1e-5)

Green points indicate novel SNPs in this suset analysis that weren't identified in the log-transformed GPSM analysis of log-transformed age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_oldage,
             prune = 0.01, 
             sig_threshold_p = 1e-5, 
             sigsnps = logage_sigsnps)+
  labs(title = "Old Animals (2011 and older): Raw Age"),
  ggmanhattan2(ran_gpsm_oldlogage,
               prune = 0.01, 
               sig_threshold_p = 1e-5, 
               sigsnps = logage_sigsnps)+
    labs(title = "Old Animals (2011 and older): Log-transformed Age"),
nrow = 2)
```

### Young Log Age

Young animal only (Born after Jan 1, 2012) Raw Age Manahattan Plots (Significance threshold -
Bonferroni)

Green points indicate novel SNPs in this transformed analysis (at
Bonferroni significance levels) that weren't identified in the GPSM
analysis of raw age.

```{r}
plot_grid(
  ggmanhattan2(ran_gpsm_younlogage,
               prune = 0.1, 
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP, 
               sigsnps2 = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & !SNP %in% logage_sigsnps) %>% 
                 .$SNP),
  ggmanhattan2(ran_gpsm_youngage,
               prune = 0.1,
               sig_threshold_p = 7.546167e-07, 
               sigsnps = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & !SNP %in% raw_age_sigsnps) %>% 
                 .$SNP, 
               sigsnps2 = filter(ran_gpsm_youngage, 
                                p < 7.546e-7 & !SNP %in% logage_sigsnps) %>% 
                 .$SNP)+
    ylim(c(0,15)),
  nrow = 2)
```


## Sequence-imputed GPSM

Used variants with MAF > 0.01 and imputation R^2 values > 0.4

During reading in I filtered p < 0.01 to make sure that R is able to handle these large files without crashing.

our Bonferroni multiple-testing cutoff is 0.05/11657021 4.289e-09

In all cases henceforth, I'm working with p-values. Calculating q-values needs the genome-wide p-values, and I'm filtering here to prevent computer from crashing. Easy enough to do this for q-values downstream though i suppose

```{r}
# ran_seq_gpsm_logage = 
#   seq(1,29) %>% 
#     purrr::map(~read_gwas(paste0("output/200910_RAN/seq_gwas/200910_RAN.log_age.chr",.x,".mlma.gz")) %>% 
#                 filter(p < 0.01)) %>% 
#     purrr::reduce(bind_rows)

ran_seq_gpsm_age = 
  seq(1,29) %>% 
    purrr::map(~read_gwas(paste0("output/200910_RAN/seq_gwas/200910_RAN.age.chr",.x,".mlma.gz")) %>% 
                 filter(p < 0.01)) %>% 
    purrr::reduce(bind_rows)


ran_gpsm_sigsnps =
  ran_seq_gpsm_age %>% 
  filter(p < 4.289e-9) %>% 
  .$SNP

ran_seq_gpsm_young_age = 
  seq(1,29) %>% 
    purrr::map(~read_gwas(paste0("output/200910_RAN/seq_gwas/200910_RAN.young_age.chr",.x,".mlma.gz")) %>% 
                 filter(p < 0.01)) %>% 
    purrr::reduce(bind_rows)

ran_seq_gpsm_old_age =
  seq(1,29) %>%
    purrr::map(~read_gwas(paste0("output/200910_RAN/seq_gwas/200910_RAN.old_age.chr",.x,".mlma.gz")) %>%
                 filter(p < 0.01)) %>%
    purrr::reduce(bind_rows)

ran_seq_gpsm_age %>% 
  filter(p < 4.289e-9) %>% 
  count()

ran_seq_gpsm_young_age %>% 
  filter(p < 4.289e-9) %>% 
  count()

ran_seq_gpsm_age %>% 
  filter(p < 5e-8) %>% 
  count()

ran_seq_gpsm_young_age %>% 
  filter(p < 5e-8) %>% 
  count()

ran_seq_gpsm_old_age_age %>% 
  filter(p < 5e-8) %>% 
  count()

```


### Manhattan Plots {.tabset}

#### Log-Transformed Age Manhattan Plot

```{r}
plot_grid(
ran_seq_gpsm_logage %>% 
  ggmanhattan2(sig_threshold_p = 4.289e-9,
               sigsnps = ran_gpsm_sigsnps),
ran_seq_gpsm_logage %>% 
  ggmanhattan2(sig_threshold_p = 4.289e-9,
               sigsnps = ran_gpsm_sigsnps)+
  ylim(c(0,15)),
nrow = 2)+
  ggtitle("Red Angus Log-Transformed Age GPSM")
```


#### Raw Age 
```{r}
plot_grid(
ran_seq_gpsm_age %>% 
  ggmanhattan2(sig_threshold_p = 4.289e-9)+
  geom_hline(yintercept = 5, color = "blue"),
ran_seq_gpsm_age %>% 
  ggmanhattan2(sig_threshold_p = 4.289e-9)+
  geom_hline(yintercept = 5, color = "blue")+
  ylim(c(0,15)),
nrow = 2)+
  ggtitle("Red Angus Log-Transformed Age GPSM")

```

#### Young Age 
```{r}
plot_grid(
  ran_seq_gpsm_young_age %>% 
    ggmanhattan2(sig_threshold_p = 4.289e-9,
               sigsnps = ran_gpsm_sigsnps)+
    geom_hline(yintercept = 5, color = "blue"),
  ran_seq_gpsm_young_age %>% 
    ggmanhattan2(sig_threshold_p = 4.289e-9,
               sigsnps = ran_gpsm_sigsnps)+
    geom_hline(yintercept = 5, color = "blue")+
    ylim(c(0,15)),
  nrow = 2)+
  ggtitle("Red Angus Young Animals (Post-2011) GPSM")

```

#### Old Age 
```{r}

ran_seq_gpsm_old_age %>% 
  ggmanhattan2(sig_threshold_p = 4.289e-9,
               sigsnps = ran_gpsm_sigsnps)+
  geom_hline(yintercept = 5, color = "blue")+
  ggtitle("Red Angus Old Animals (Pre-2012) GPSM")

```
#### Single Chromosome

```{r}
single_chrom = read_gwas("output/200910_RAN/seq_gwas/200910_RAN.age.chr6.mlma.gz") %>% 
  mutate(COJO = case_when(SNP %in% ran_age_seq_cojo$SNP ~ TRUE,
                          TRUE ~ FALSE),
         BP = BP/1000000)


plot_grid(
    single_chrom %>% 
      # filter(BP > 8e7,
      #        BP <10e7) %>% 
    ggmanhattan2(prune = 0.1,
                 sigsnps = single_chrom %>% 
                   filter(COJO == TRUE) %>% 
                   .$SNP,
                 sig_threshold_p = 5e-8)+
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10))+
  labs(x = "Megabases"),
  ran_nsl %>% 
    filter(CHR == 6) %>% 
    selscan_manhattans(col = abs(norm_nsl))+
    geom_vline(xintercept = single_chrom %>% 
                   filter(COJO == TRUE) %>% 
                   .$BP, 
               color = "springgreen3",
               alpha = 0.5)+
    labs(y = "Normalized nSL", x = "Megabases")+
    geom_hline(yintercept = 3, color = "red")+
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)),
  ran_xpnsl %>% 
    mutate(BP = BP/1000000) %>% 
    filter(CHR == 6,
           abs(norm_xpnsl) > 1) %>% 
    selscan_manhattans(col = abs(norm_xpnsl))+
    geom_vline(xintercept = single_chrom %>% 
                   filter(COJO == TRUE) %>% 
                   .$BP, 
               color = "springgreen3",
               alpha = 0.5)+
    labs(y = "RAiSD Mu", x = "Megabases")+
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)),
  ran_raisd %>% 
    mutate(BP = BP/1000000) %>% 
    filter(CHR == 6,
           Mu > 50) %>% 
    selscan_manhattans(col = Mu)+
    geom_vline(xintercept = single_chrom %>% 
                   filter(COJO == TRUE) %>% 
                   .$BP, 
               color = "springgreen3",
               alpha = 0.5)+
    labs(y = "RAiSD Mu", x = "Megabases")+
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)),
  nrow = 3)

ran_nsl =
  read_tsv("output/200910_RAN/selscan/nsl/200910_RAN.all.nsl.out.100bins.norm.gz",
             col_names = c("SNP", "BP", "freq", "sl1", "sl0", "nsl", "norm_nsl", "sig")) %>% 
    mutate(CHR = as.numeric(str_split_fixed(SNP, ":", n = 4)[,1]),
           SNP = paste(CHR, BP, sep = ":"),
         sig = case_when(abs(norm_nsl) > 3 ~ TRUE,
                         TRUE ~ FALSE),
         BP = BP/1000000)


single_chrom_manhattan(single_chrom, prune = 1)
  
single_chrom_manhattan(prune = 1)


ran_nsl_window = 
  read_tsv("output/200910_RAN/selscan/nsl/200910_RAN.all.nsl.windows.txt") %>% 
  mutate(nsl_win = abs(Wstat),
         SNP = paste(CHR, BP, sep = ":")) 
  
```


## Sequence-imputed COJO


[COJO documentation](https://cnsgenomics.com/software/gcta/#COJO)

[COJO citation](https://www.nature.com/articles/ng.2213)

Used `--cojo-slct` procedure with a significance cutoff of 5e-8

Reading in single-chromosome COJO files.
```{r}
ran_age_seq_cojo = 
  list.files(path='output/200910_RAN/seq_cojo/',pattern="200910_RAN.age.chr") %>% 
  purrr::map(
  ~read_tsv(paste0("output/200910_RAN/seq_cojo/",.x),
           col_names = c("Chr", "SNP", "bp", "freq", "refA", "b", "se", "p", "n", "freq_geno", "bJ", "bJ_se", "pJ", "LD_r"),
           skip = 1) %>% 
    select(SNP, CHR = Chr, BP = bp, p = pJ)) %>% 
    purrr::reduce(bind_rows) 

ran_old_seq_cojo = 
  list.files(path='output/200910_RAN/seq_cojo/',pattern="200910_RAN.old_age") %>% 
  purrr::map(
  ~read_tsv(paste0("output/200910_RAN/seq_cojo/",.x),
           col_names = c("Chr", "SNP", "bp", "freq", "refA", "b", "se", "p", "n", "freq_geno", "bJ", "bJ_se", "pJ", "LD_r"),
           skip = 1) %>% 
    select(SNP, CHR = Chr, BP = bp, p = pJ)) %>% 
    purrr::reduce(bind_rows) 

ran_young_seq_cojo = 
  list.files(path='output/200910_RAN/seq_cojo/',pattern="200910_RAN.young_age.chr") %>% 
  purrr::map(
  ~read_tsv(paste0("output/200910_RAN/seq_cojo/",.x),
           col_names = c("Chr", "SNP", "bp", "freq", "refA", "b", "se", "p", "n", "freq_geno", "bJ", "bJ_se", "pJ", "LD_r"),
           skip = 1) %>% 
    select(SNP, CHR = Chr, BP = bp, p = pJ)) %>% 
    purrr::reduce(bind_rows)

```

### FAETH Score Annotations for COJO Hits
1) Write out SNP names, will just use grep in `joinable_FAETH_ARS.txt` file to find these SNPs as reading in isn't exactly helpful
2) Read them in and compare to distribution of FAETH scores for each 
Commands:

```{bash}
> grep "1:720042 \|1:773427 \|1:1071701 \|1:1134753 \|1:1569696 \|1:2268272 \|1:9037085 \|1:148638095 \|10:509620 \|10:2377878 \|10:3304349 \|10:7702688 \|10:77760348 \|10:84326255 \|10:89215226 \|11:7233185 \|11:8355973 \|11:9227108 \|11:10908641 \|11:11211958 \|11:11399548 \|11:13399691 \|11:17113988 \|11:17539086 \|11:17890138 \|11:18921361 \|11:20317517 \|11:20480945 \|11:21202326 \|11:21938093 \|11:22625682 \|11:36448878 \|11:41409775 \|11:43784961 \|11:45947339 \|11:47469539 \|11:54790754 \|11:59101452 \|11:59119877 \|13:4743414 \|13:5262550 \|13:11822579 \|13:36339812 \|13:37848629 \|14:22518 \|14:156254 \|14:1868236 \|14:4050892 \|14:4732529 \|14:5453303 \|14:11886077 \|14:12258158 \|14:23257180 \|14:24451792 \|16:21295945 \|16:68631334 \|18:690712 \|18:9497918 \|18:14354917 \|18:17213715 \|18:18535148 \|18:19227993 \|18:19529870 \|18:27140565 \|18:27490963 \|18:28500703 \|18:29125466 \|18:29216351 \|18:30026611 \|18:30134779 \|19:412779 \|19:1016661 \|19:1391724 \|19:2361572 \|19:3490448 \|19:3507526 \|19:3804333 \|19:3944451 \|19:3972539 \|19:9332861 \|19:10952651 \|19:11003839 \|19:13728353 \|19:15955031 \|19:17335586 \|19:21319644 \|19:21326628 \|19:25755143 \|19:29454421 \|2:6757167 \|2:7498211 \|2:8100524 \|2:21860593 \|2:27120579 \|2:111403008 \|2:113671290 \|2:127446835 \|20:30112844 \|20:65179823 \|21:1148483 \|21:2117530 \|21:3445763 \|21:4075157 \|21:5237404 \|21:6722575 \|21:7103453 \|21:8187835 \|21:11084133 \|21:11146608 \|21:11158491 \|21:11234701 \|21:12085920 \|21:12117079 \|21:12148665 \|21:14130322 \|21:14552611 \|21:24117786 \|21:38865360 \|21:39317383 \|21:62593095 \|23:1032665 \|23:1324051 \|23:1366335 \|23:1366691 \|23:1768070 \|23:1783688 \|23:1805412 \|23:2449698 \|23:2541461 \|23:2599415 \|23:3224854 \|23:3303533 \|23:3932992 \|23:3948262 \|23:5203364 \|23:5843917 \|23:9888981 \|23:12050166 \|23:12276093 \|23:12541030 \|23:14714998 \|23:14790825 \|23:14817774 \|23:15306345 \|23:20796813 \|23:23147472 \|23:48398507 \|23:48455109 \|23:50442856 \|25:8955090 \|25:9065380 \|26:23082541 \|26:29784658 \|26:29979085 \|26:31429344 \|26:32335380 \|26:32357492 \|26:32401666 \|26:33408375 \|26:34425381 \|26:34697940 \|26:34988139 \|26:41876216 \|26:42049644 \|26:42553807 \|26:42811821 \|27:405898 \|28:31307728 \|28:32191104 \|28:33441407 \|28:34745056 \|28:42523583 \|28:43687081 \|3:86398399 \|3:90423016 \|3:90841317 \|3:91361195 \|3:92833758 \|3:93064646 \|3:95714594 \|3:97932705 \|3:98861495 \|3:101426669 \|3:112272306 \|4:4055714 \|5:85176900 \|5:85398036 \|6:11410294 \|6:12842123 \|6:14196268 \|6:18234949 \|6:19085539 \|6:19831619 \|6:21771162 \|6:22420738 \|6:23196562 \|6:25861907 \|6:26304356 \|6:26453659 \|6:30869913 \|6:31848172 \|6:35947056 \|6:36059193 \|6:36072398 \|6:38546883 \|6:44918748 \|6:48898755 \|6:49841928 \|6:56961317 \|6:59577310 \|6:60189953 \|6:114937887 \|7:26806870 \|7:53177641 \|7:53888447 \|7:56681778 \|7:58876521 \|7:64951198 \|7:78069361 \|7:79332259 \|7:84836277 \|7:88023501 \|7:88102065 \|7:88114202 \|7:88897740 \|7:89137291 \|7:89895056 \|7:93134879 \|7:99349462 \|7:99883674 \|7:101383629 \|7:107668986 \|8:69385 \|9:846219 \|9:10086241 \|9:15850396 \|9:18052071 \|9:37723214 \|9:37773816 \|9:38446163 \|9:39591442 \|9:46641263 \|9:49231978 \|9:54937793 \|9:55526655 \|9:91100383 \|9:95146355 \|9:101795581 " data/FAETH_intermediates/joinable_FAETH_ARS.txt > output/200910_RAN/seq_cojo/FAETH/200910_RAN.age.cojo.FAETH.txt

> grep "1:722538 \|1:773427 \|1:1052361 \|1:1063897 \|1:1071701 \|1:1134753 \|1:1896654 \|1:2268272 \|1:6256234 \|1:6612273 \|1:8083887 \|1:9010333 \|1:9475931 \|1:10203552 \|1:10604349 \|1:10766013 \|1:10913893 \|1:10940482 \|1:11054832 \|1:16852592 \|1:18713212 \|1:19560010 \|1:20635456 \|10:815280 \|10:2331204 \|10:3316433 \|10:5423142 \|10:7702688 \|10:16175889 \|10:16404552 \|10:88723483 \|10:89215226 \|11:7311711 \|11:7621506 \|11:8317677 \|11:8515583 \|11:9936603 \|11:10430187 \|11:10634416 \|11:10856174 \|11:11221016 \|11:11362173 \|11:14097199 \|11:15485650 \|11:17056530 \|11:17476804 \|11:17573211 \|11:17592781 \|11:17592868 \|11:17692678 \|11:17849450 \|11:17890604 \|11:18915325 \|11:19965519 \|11:19973793 \|11:20569541 \|11:28679160 \|11:29174803 \|11:69399750 \|12:442568 \|12:688001 \|12:808457 \|12:840165 \|12:863199 \|12:2604286 \|12:9647346 \|12:9959897 \|12:10421924 \|12:10502044 \|12:10533569 \|12:12559315 \|12:12989521 \|12:13206923 \|12:45173652 \|12:46730507 \|12:51070398 \|12:55506324 \|12:61066020 \|12:65137647 \|12:68013938 \|13:1113875 \|13:6027097 \|13:11822579 \|13:37386896 \|13:37625849 \|13:37848629 \|13:46808594 \|13:47262326 \|13:47673290 \|13:48970065 \|13:50304849 \|13:58381692 \|13:58917842 \|13:65286896 \|14:22518 \|14:156254 \|14:7976475 \|14:8506714 \|14:17390849 \|14:18387623 \|14:23001109 \|14:23257180 \|14:23592917 \|14:24452175 \|14:28335410 \|14:29900938 \|14:33178767 \|14:33669060 \|14:39327333 \|14:47307478 \|14:47808662 \|14:49115171 \|14:50403790 \|14:51006965 \|14:54434822 \|14:55152864 \|14:55639734 \|14:57977938 \|14:58111264 \|14:59047489 \|14:59747342 \|14:59788659 \|14:60150582 \|14:60322029 \|14:60456303 \|14:60799025 \|14:60831844 \|14:61585567 \|14:62971728 \|14:68145523 \|14:77585862 \|14:78036034 \|14:78037164 \|14:78044125 \|14:79172630 \|15:44987136 \|15:47321319 \|15:47481607 \|15:57165660 \|15:57383227 \|15:64647508 \|15:64765703 \|15:64781739 \|15:74470423 \|15:74538358 \|15:74753121 \|15:76883927 \|16:957211 \|16:1047025 \|16:1070414 \|16:1374913 \|16:11069751 \|16:71418875 \|16:73942705 \|17:51861826 \|18:63778412 \|18:64748758 \|19:242064 \|19:412779 \|19:1099601 \|19:1100413 \|19:1204420 \|19:1783988 \|19:2606317 \|19:3827679 \|19:3944451 \|19:4200195 \|19:4253734 \|19:6028929 \|19:10870320 \|19:16911488 \|19:25798935 \|2:7500278 \|2:8100524 \|2:8674849 \|2:9826292 \|2:10679762 \|2:12194425 \|2:12208106 \|2:12363109 \|2:12386721 \|2:12732917 \|2:15449766 \|2:15764320 \|2:16047868 \|2:19635098 \|2:22112854 \|2:22339424 \|2:22745800 \|2:22868999 \|2:23027621 \|2:23259680 \|2:28639397 \|2:28823286 \|2:30374233 \|2:32952841 \|2:42247695 \|2:45247236 \|2:49303806 \|2:53006426 \|2:53126236 \|2:53151542 \|2:53189373 \|2:53283910 \|2:53353546 \|2:53390590 \|2:53730042 \|2:53803504 \|2:60054014 \|2:63289368 \|2:68387467 \|2:76483168 \|2:92863381 \|2:102818589 \|2:103217345 \|2:103225507 \|2:103741954 \|2:104899441 \|2:105543115 \|2:113229961 \|2:136159490 \|20:49419382 \|21:1146167 \|21:1778727 \|21:2117530 \|21:3728703 \|21:3743512 \|21:3768317 \|21:3887410 \|21:4642063 \|21:4806372 \|21:7081229 \|21:7103223 \|21:8170292 \|21:8971599 \|21:10755053 \|21:11029235 \|21:11132658 \|21:11154544 \|21:11177811 \|21:11244178 \|21:12976221 \|21:64875723 \|21:64957035 \|23:490909 \|23:838197 \|23:1324051 \|23:1366335 \|23:1366691 \|23:1768070 \|23:1800642 \|23:1832160 \|23:1918909 \|23:2082004 \|23:2599415 \|23:2654697 \|23:4387101 \|23:5113023 \|23:10411911 \|23:10543086 \|23:11354737 \|23:11747930 \|23:11777742 \|23:11805518 \|23:12616219 \|23:13129316 \|23:20037462 \|24:17810958 \|24:25493847 \|24:27206537 \|24:27780789 \|24:27966765 \|24:28893957 \|24:36456552 \|24:36955648 \|24:37571966 \|24:38012259 \|24:38947548 \|24:38947749 \|24:38949876 \|24:38968511 \|24:38971013 \|24:39021058 \|24:39420660 \|24:40634009 \|24:43951516 \|24:49301748 \|24:51458941 \|25:8738478 \|25:9065380 \|25:36599479 \|26:31210993 \|26:32346219 \|26:32357492 \|26:32385055 \|26:32641874 \|26:34205775 \|26:34432196 \|26:34558428 \|26:34735344 \|26:34988139 \|26:35126262 \|26:35333465 \|26:35983568 \|26:36422482 \|26:37875268 \|26:38112908 \|26:42353937 \|26:43954288 \|26:45137144 \|26:45272465 \|26:45329895 \|26:45451492 \|26:49325027 \|26:49432811 \|28:613012 \|3:7808113 \|3:13334968 \|3:13344300 \|3:13394995 \|3:13977241 \|3:17251848 \|3:23902971 \|3:24118166 \|3:25087305 \|3:64252547 \|3:86981654 \|3:87440743 \|3:87835375 \|3:88313479 \|3:91361195 \|3:91481076 \|3:92813097 \|3:98011212 \|3:98946372 \|3:102034343 \|3:117849750 \|3:118641997 \|4:50093574 \|4:50134171 \|4:50931071 \|4:50979242 \|5:53033839 \|6:599233 \|6:842877 \|6:949244 \|6:28677939 \|6:35960688 \|6:39471772 \|6:39499785 \|6:39824376 \|6:41304182 \|6:43058482 \|6:47075645 \|6:47510463 \|6:48908886 \|6:49355845 \|6:49435958 \|6:49794516 \|6:49826827 \|6:49885440 \|6:49933549 \|6:50776379 \|6:53490245 \|6:59811218 \|6:89631901 \|6:104267459 \|6:105011885 \|6:113860654 \|6:114750937 \|6:114921606 \|6:114961357 \|6:115122390 \|6:115236183 \|6:115339089 \|6:115949269 \|6:116003987 \|7:358667 \|7:369503 \|7:497178 \|7:599670 \|7:642673 \|7:2121310 \|7:10388023 \|7:10393838 \|7:12089146 \|7:12371154 \|7:88800495 \|7:88803806 \|7:89375787 \|9:5319237 \|9:14601128 \|9:15631441 \|9:15770551 \|9:15850396 \|9:15857514 \|9:23244267 \|9:27337347 \|9:32139049 \|9:32936480 \|9:36012154 \|9:36560042 \|9:37723231 \|9:39413842 \|9:49045328 \|9:79713850 \|9:80367390 \|9:81961273 \|9:82349415 \|9:82703071 \|9:84854163 \|9:91319725 \|9:91606673 \|9:92104822 \|9:92753867 \|9:94383273 \|9:96112707" data/FAETH_intermediates/joinable_FAETH_ARS.txt > output/200910_RAN/seq_cojo/FAETH/200910_RAN.young_age.cojo.FAETH.txt
```

The top 1/2 FAETH score cutoff = 1.48e-8
The top 1/3 FAETH score cutoff = 1.60e-8
```{r}
FAETH_Scores = 
  vroom("data/FAETH.txt.gz",
        delim = "\t") %>% 
  filter(!is.na(FAETH))

ran_age_seq_cojo_FAETH = 
  read_delim("output/200910_RAN/seq_cojo/200910_RAN.age.cojo.FAETH.txt",
           col_names = c("Chr", "bp", "SNP", "chr", "ARS_BP", "bull.FAETH1.1", "cow.FAETH1.1", "annotation", "aseQTL", "eeQTL", "geQTL", "bull.ldscore50kb", "bull.maf", "ChIPSeq", "mQTL", "selection.sig", "young", "variant.density50kb", "sQTL", "conserved100way", "cow.ldscore50kb", "cow.maf"), 
           col_type = cols(SNP = col_character()),
           delim = " ") %>% 
  mutate(FAETH = (bull.FAETH1.1 + cow.FAETH1.1)/2,
         top_half = case_when(FAETH >= 1.48e-8 ~ TRUE,
                              TRUE ~ FALSE),
         top_third = case_when(FAETH >= 1.60e-8 ~ TRUE,
                              TRUE ~ FALSE))

ran_young_seq_cojo_FAETH = 
  read_delim("output/200910_RAN/seq_cojo/200910_RAN.young_age.cojo.FAETH.txt",
           col_names = c("Chr", "bp", "SNP", "chr", "ARS_BP", "bull.FAETH1.1", "cow.FAETH1.1", "annotation", "aseQTL", "eeQTL", "geQTL", "bull.ldscore50kb", "bull.maf", "ChIPSeq", "mQTL", "selection.sig", "young", "variant.density50kb", "sQTL", "conserved100way", "cow.ldscore50kb", "cow.maf"),
           col_type = cols(SNP = col_character()),
           delim = " ") %>% 
  mutate(FAETH = (bull.FAETH1.1 + cow.FAETH1.1)/2,
         top_half = case_when(FAETH >= 1.48e-8 ~ TRUE,
                              TRUE ~ FALSE),
         top_third = case_when(FAETH >= 1.60e-8 ~ TRUE,
                              TRUE ~ FALSE))


ran_age_seq_cojo_FAETH %>% count(top_half)
ran_age_seq_cojo_FAETH %>% count(top_third)
ran_young_seq_cojo_FAETH %>% count(top_half)
ran_young_seq_cojo_FAETH %>% count(top_third)
```



### Gene Annotations for COJO hits:

Annotated genes within 50 kb of COJO SNP

#### Age Genes
```{r}
ran_age_seq_cojo %>% 
  gene_annotation() %>% 
    left_join(ran_age_seq_cojo, 
              by = c("CHR", "BP")) %>% 
  select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>%
  datatable()

ran_age_seq_cojo %>% 
  gene_annotation() %>% 
    right_join(ran_age_seq_cojo, 
              by = c("CHR", "BP")) %>%
  mutate(ingene = case_when(distance == 0 ~ "within",
                            is.na(distance) ~ ">50kb",
                            TRUE ~ "proximal")) %>% 
  group_by(SNP) %>% 
  top_n(1, wt = -distance) %>% 
  ungroup() %>% 
  count(ingene)

ran_young_seq_cojo %>% 
  gene_annotation() %>% 
    right_join(ran_young_seq_cojo, 
              by = c("CHR", "BP")) %>%
  mutate(ingene = case_when(distance == 0 ~ "within",
                            is.na(distance) ~ ">50kb",
                            TRUE ~ "proximal")) %>% 
  group_by(SNP) %>% 
  top_n(1, wt = -distance) %>% 
  ungroup() %>% 
  count(ingene)

```
### Shared domains with selective sweep meta-analysis

#### Age
```{r}
ran_age_seq_cojo_genes =
  ran_age_seq_cojo %>% 
    gene_annotation() %>% 
      left_join(ran_age_seq_cojo, 
                by = c("CHR", "BP")) %>% 
    select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% 
  select(gene_name) %>% 
  distinct() %>% 
  .$gene_name

sweeps = read_csv("data/meata_sweeeps.csv")

ran_age_seq_cojo_genes %>% 
  purrr::map(
    ~filter(sweeps,str_detect(`Gene Names`, .x)) %>% 
      mutate(driver = .x)
    ) %>% 
  reduce(bind_rows) %>% 
  count(driver)
   
    
```

#### Young Age
```{r}
ran_young_seq_cojo_genes =
  ran_young_seq_cojo %>% 
    gene_annotation() %>% 
      left_join(ran_young_seq_cojo, 
                by = c("CHR", "BP")) %>% 
    select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% View()
  select(gene_name) %>% 
  distinct() %>% 
  .$gene_name

sweeps = read_csv("data/meata_sweeeps.csv")
  
small_sweeps = read_csv("data/meata_sweeeps.csv") %>% 
  mutate(size = as.numeric(str_split_fixed(`Validated MSS region (Mb)`, pattern = "-", n = 2)[,2]) - as.numeric(str_split_fixed(`Validated MSS region (Mb)`, pattern = "-", n = 2)[,1])) %>% 
  filter(size < 2.5)

ran_young_seq_cojo_genes %>% 
  purrr::map(
    ~filter(sweeps,str_detect(`Gene Names`, .x)) %>% 
      mutate(driver = .x)
    ) %>% 
  reduce(bind_rows) %>% 
  count(driver)

ran_young_seq_cojo_genes %>% 
  purrr::map(
    ~filter(small_sweeps,str_detect(`Gene Names`, .x)) %>% 
      mutate(driver = .x)
    ) %>% 
  reduce(bind_rows) %>% 
  count(driver)
   
sweeps %>% 
  ggplot(aes(size))+
  geom_histogram()
```

### Gene Ontology Analysis

Gene ontology enrichment analysis using `gprofiler2::gost()` of gene names from GALLO annotation of log-transformed GPSM COJO hits

Significance threshold set at FDR < 0.10
```{r, eval }
ran_age_seq_cojo_GO =
  ran_age_seq_cojo %>% 
  gene_annotation(search_bp = 50000) %>% 
    left_join(ran_age_seq_cojo, 
              by = c("CHR", "BP")) %>% 
  select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% 
  .$gene_name %>% unique() %>% 
  gost(query = ., 
      organism = "btaurus",
      user_threshold = 0.1,
      correction_method = "fdr")


gostplot(ran_age_seq_cojo_GO, capped = TRUE, interactive = TRUE)
```

#### Young Age
```{r}
ran_young_seq_cojo %>% 
  gene_annotation() %>% 
    left_join(ran_young_seq_cojo, 
              by = c("CHR", "BP")) %>% 
  select(-chr, -start_pos, -end_pos, -gene_biotype, -width, -strand, -upstream, -downstream) %>% 
  datatable()
```

### QTL Enrichments {.tabset}

Results are for Red Angus sequence-based COJO hits with p < 1e-5. QTL are within 50 kb of identified SNP

Removed milk-associated QTL, perform enrichment on whole genome QTL and by chromosome

#### Age (Genome)
```{r}
ran_age_seq_cojo %>%
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```

#### Young (Genome)
```{r}
ran_young_seq_cojo %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```

#### Age (per-Chromosome)
```{r}
ran_age_seq_cojo %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```

#### Young (per-chromosme)
```{r}
ran_young_seq_cojo %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```





## Conditional Joint Analysis (COJO)


### Candidate genes associated with COJO hits:

Using GALLO annotation to identify nearby genes (within 50 kb) to significant (p < 1e-5) COJO hits

#### Log Age
```{r}

ran_gpsm_log_age_cojo = 
  read_tsv("output/200910_RAN/cojo/200910_RAN.log_age.850K.jma.cojo.gz") %>% 
    select(CHR = Chr, BP = bp, p = pJ) %>% 
    gene_annotation() %>% 
    left_join(read_tsv("output/200910_RAN/cojo/200910_RAN.log_age.850K.jma.cojo.gz"), 
              by = c("CHR" = "Chr", "BP" = "bp")) %>% 
  select(-chr, -start_pos, -end_pos, -SNP, -gene_biotype, -width, -strand, -upstream, -downstream, -refA, -b, -se, -freq_geno, -n)
dim(ran_gpsm_log_age_cojo)

datatable(ran_gpsm_age_cojo)
```

#### Raw Age
```{r}

ran_gpsm_age_cojo = 
  read_tsv("output/200910_RAN/cojo/200910_RAN.age.850K.jma.cojo.gz") %>% 
    select(CHR = Chr, BP = bp, p = pJ) %>% 
    gene_annotation() %>% 
    left_join(read_tsv("output/200910_RAN/cojo/200910_RAN.age.850K.jma.cojo.gz"), 
              by = c("CHR" = "Chr", "BP" = "bp")) %>% 
  select(-chr, -start_pos, -end_pos, -SNP, -gene_biotype, -width, -strand, -upstream, -downstream, -refA, -b, -se, -freq_geno, -n)
dim(ran_gpsm_age_cojo)

datatable(ran_gpsm_age_cojo)
```


### Gene Ontology Analysis

Gene ontology enrichment analysis using `gprofiler2::gost()` of gene names from GALLO annotation of log-transformed GPSM COJO hits

Significance threshold set at FDR < 0.10
```{r}
GO_results = gost(query = ran_gpsm_age_cojo$gene_name, 
                  organism = "btaurus",
                  user_threshold = 0.1,
                  correction_method = "fdr")


gostplot(GO_results, capped = TRUE, interactive = TRUE)
```

### QTL-enrichment analysis (within annotated QTL)

These QTL are quite small (only 4 bp), things certainly change when we search for a larger area around them.
```{r}
read_tsv("output/200910_RAN/cojo/200910_RAN.age.850K.jma.cojo.gz") %>% 
  select(CHR = Chr, BP = bp, p = pJ) %>% 
  qtl_annotation(search_bp = 0) %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```
### QTL-enrichment analysis (within 50kb annotated QTL)

These QTL are quite small (only 4 bp) we identify much many more when we're searching in a substantial area around them.
```{r}
read_tsv("output/200910_RAN/cojo/200910_RAN.age.850K.jma.cojo.gz") %>% 
  select(CHR = Chr, BP = bp, p = pJ) %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "genome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```

### Per-Chromosome QTL-enrichment analysis (within 50kb of significant SNP)
```{r}
read_tsv("output/200910_RAN/cojo/200910_RAN.age.850K.jma.cojo.gz") %>% 
  select(CHR = Chr, BP = bp, p = pJ) %>% 
  qtl_annotation() %>% 
  filter(QTL_type != "Milk") %>% 
  qtl_enrich(qtl_db = qtl, qtl_file = ., qtl_type = "Name", enrich_type = "chromosome", nThreads = 2) %>% 
  arrange(adj.pval) %>% 
  datatable()
```



